Data table  {data-orientation=rows}
================================================================================


Inputs {.sidebar data-width=250}
--------------------------------------------------------------------------------
<br>
**Data Table.** Descriptive statistical analysis on a gene level with visualisations. 

```{r}
##----------------------------------------------------------------------
## More Info (pop up window)
##----------------------------------------------------------------------
tipify( actionLink("data_info", "Info", icon = icon("info-circle")),
       "Show more information about this module.")

dropdown_search_gene='<code>Search gene</code>'
menu_grouped='<code>grouped</code>'
menu_options='<code>Options</code>'

data_infotext = paste0('The <strong>data table</strong> module provides a descriptive statistical analysis on a gene level with visualisations. <br><br> Users can start the analysis by selecting a gene of interest in the dropdown menu ',dropdown_search_gene,'. <br><br> During the visual analysis, users can choose multiple options from the ',menu_options,'. They can filter out some samples or collapse the samples by predetermined groups by clicking ',menu_grouped,'. It is also possible to visualize the information on a raw count level, count per million (CPM), or rather on a logarithmic expression level (logCPM). <br><br> To download the raw counts or the results of differential gene expression and gene set enrichment analysis, please navigate to the <code>Download</code> links in the left-bottom.')

observeEvent( input$data_info, {
    showModal(modalDialog(
        title = HTML("<strong>Data table Module</strong>"),
        HTML(data_infotext),
        easyClose = TRUE ))
})
```


```{r}
require(htmltools)
## data set parameters
##textInput("search_gene","Filter genes", value="")

br();br();br();br();
tipify( selectInput("search_gene","Search gene", choices=NULL),
       "Select a gene of interest for the analysis.", placement="top")

br();br();
tipify( actionLink("dt_options", "Options", icon=icon("cog", lib = "glyphicon")),
       "Toggle advanced options.", placement="top");br()
##actionLink("dt_options2a", "Options ...", icon=icon("cog", lib = "glyphicon"));br()
##actionLink("dt_options2c", "Options ...", icon=icon("cogs", lib = "fontawesome"));br()

## br();br()
datatypes <- c("CPM","logCPM")
if(PRO.VERSION) datatypes <- c("counts","CPM","logCPM")
conditionalPanel(
    "input.dt_options % 2 == 1",
    tagList(
        tipify( selectInput("data_samplefilter","Filter samples:", choices=NULL, multiple=TRUE),
                "Filter unwanted samples out of the analysis.", placement="top"),
        tipify( checkboxInput('data_grouped','grouped',TRUE),
                "Group the samples by predefined phenotype.", placement="top"),
        tipify( radioButtons('data_type','Data type:', choices=datatypes, selected="logCPM", inline=TRUE),
                "Choose an input data type for the analysis.", placement="bottom"),

        br(),HTML("Download tables: "),br(),
        tipify( downloadLink('data_downloadData', '[counts]'),
                "Download the raw counts.", placement="bottom"),
        tipify( downloadLink('gx_downloadData', ' [comparisons]'),
                "Download the results of differential gene expression analysis.", placement="bottom"),
        tipify( downloadLink('gs_downloadData', ' [enrichment]'),
                "Download the results of enrichment analysis.", placement="bottom")
    )
)


## dropdownButton(
##     downloadLink('data_downloadData', '[counts]'),
##     downloadLink('gx_downloadData', ' [comparisons]'),
##     downloadLink('gs_downloadData', ' [enrichment]'),
##     circle = TRUE, size = "xs", ## status = "danger",
##     icon = icon("download"), width = "80px",
##     tooltip = tooltipOptions(title = "Download", placement = "right")
## )

##br();br();br();br();br();br()

## ------- download functions -----------
output$data_downloadData <- downloadHandler(
    filename = function() {
        paste("playground-COUNTS.csv", sep="")
    },
    content = function(file) {
        ngs <- inputData()
	gg <- ngs$genes[rownames(ngs$counts),]
	if(class(gg)=="DataFrame") gg <- do.call(cbind, gg@listData)
        df <- data.frame(gg, round(ngs$counts,digits=5))
        write.csv(df, file, row.names=FALSE)
    }
)

output$gx_downloadData <- downloadHandler(
    filename = function() {
        ##ngs <- inputData()
        paste("playground-COMPARISONS-all.csv", sep="")
    },
    content = function(file) {
        ngs <- inputData()
        mx <- NULL
        k=1
        for(k in names(ngs$gx.meta$meta)) {
            colnames(ngs$gx.meta$meta[[k]])
            ff <- ngs$gx.meta$meta[[k]][,c("meta.fx","meta.p","meta.q")]
            colnames(ff) <- paste(c("logFC","p","q"),k,sep=".")
            if(!is.null(mx)) {
                mx <- cbind(mx, ff)
            } else {
                mx <- ff
            }
        }
        mx <- mx[,order(colnames(mx))]
        pp <- rownames(ngs$gx.meta$meta[[1]])
	gg <- ngs$genes[pp,]
	if(class(gg)=="DataFrame") gg <- do.call(cbind, gg@listData)
        mx <- data.frame(gg, mx)
        write.csv(mx, file, row.names=FALSE)
    }
)

output$gs_downloadData <- downloadHandler(
    filename = function() {
        ##ngs <- inputData()
        paste("playground-ENRICHMENT-all.csv", sep="")
    },
    content = function(file) {
        ngs <- inputData()
        mx <- NULL
        for(k in names(ngs$gset.meta$meta)) {
            ff <- ngs$gset.meta$meta[[k]][,c("meta.fx","meta.q")]
            colnames(ff) <- paste(c("logFC","q"),k,sep=".")
            if(!is.null(mx)) {
                mx <- cbind(mx, ff)
            } else {
                mx <- ff
            }
        }
        mx <- mx[,order(colnames(mx))]
        aa <- ngs$gset.meta$meta[[1]][,c("gene.set")]
        mx <- cbind( gene.set=aa, mx)
        write.csv(mx, file, row.names=FALSE)
    }
)

## ------- observe functions -----------

## update filter choices upon change of data set 
observe({
    ngs <- inputData()
    if(is.null(ngs)) return(NULL)

    ## levels for sample filter
    levels = getLevels(ngs$Y)
    updateSelectInput(session, "data_samplefilter", choices=levels)
    genes <- sort(ngs$genes[rownames(ngs$X),]$gene_name)
    sel = genes[1]  ## most var gene
    updateSelectInput(session,'search_gene', choices=genes, selected=sel)
})

```


Col {.tabset data-height=400}
--------------------------------------------------------------------------------

### Plots

```{r echo=FALSE, warning=FALSE, fig.height=5, fig.width=10}

##----------------------------------------------------------------------
##                     Info messsages
##----------------------------------------------------------------------

OMIM="<a href='https://www.ncbi.nlm.nih.gov/omim/'> OMIM</a>"
KEGG="<a href='https://www.ncbi.nlm.nih.gov/pmc/articles/PMC102409/'> KEGG</a>"
GO="<a href='http://geneontology.org/'> GO</a>"
ImmProt="<a href='https://www.ncbi.nlm.nih.gov/pubmed/28263321'> ImmProt</a>"
HPA="<a href='https://www.nature.com/articles/nbt1210-1248'> HPA</a>"
GTEx="<a href='https://www.ncbi.nlm.nih.gov/pubmed/23715323'> GTEx</a>"

data_genePlots_tsne_text=paste0('A <b>t-SNE clustering</b> of samples (or cells) colored by an expression of a gene selected by users in the ',dropdown_search_gene, ' dropdown menu. The red color represents an over-expression of the selected gene across samples (or cells).')
data_genePlots_barboxplot_text=paste0('An expression barplot of grouped samples (or cells) for a gene specified by the user in the ',dropdown_search_gene,'. The samles (or cells) in the barplot can be ungrouped by unclicking the ',menu_grouped, ' in the main ',menu_options, '.')
data_genePlots_correlationplot_text=paste0('The top <em>N</em>={16} positively and negatively correlated genes barplot with a gene specified by the user in the ',dropdown_search_gene,' across samples (or cells). The expression levels of genes are colored in the barplot, where the low and high expressions range between the light grey and dark black colors, respectively.')
data_genePlots_waterfallplot_text=paste0('An average expression ranking of a gene specified by the user in the ',dropdown_search_gene,' in  currently selected dataset.')
data_geneInfo_text = paste0('To find out more information from the literature, hyperlinks are provide to connect the selected gene in the  ',dropdown_search_gene,' to public databases, including ', OMIM,', ', KEGG, ' and ',GO,'.')
data_corplot_text = paste0('The top <em>N</em>={20} cumulative positively and negatively correlated genes barplot with a gene specified by the user in the ',dropdown_search_gene,' across samples (or cells) in currently selected dataset as well as in public datasets such as ',ImmProt,' and ',HPA,'. The correlations of genes for each dataset are respectively colored in the barplot.')
data_tissueplot_text = paste0('A tissue expression barplot for a selected gene by the user in the ',dropdown_search_gene,' using the genotype-tissue expression (',GTEx,') dataset.')


##----------------------------------------------------------------------
##                     Waterfall plot
##----------------------------------------------------------------------

data_genePlots_waterfallplotFUNC <- reactive({
    require(RColorBrewer)
    
    ngs <- inputData()
    if(is.null(ngs)) return(NULL)
    
    gene = "KCNN4"
    gene = ngs$genes$gene_name[1]
    if(!is.null(input$search_gene) && input$search_gene!="") gene <- input$search_gene
    samples = colnames(ngs$X)
    if(!is.null(input$data_samplefilter)) {
        samples <- selectSamplesFromSelectedLevels(ngs$Y, input$data_samplefilter)
    }
    nsamples = length(samples)

    par(mar=c(10,3.5,3,1), mgp=c(2.1,0.8,0))

    if(input$data_type=="counts") {
        mean.fc <- sort(rowMeans(ngs$counts[,samples,drop=FALSE]),decreasing=TRUE)
        ylab="expression (counts)"
    }
    if(input$data_type=="CPM") {
        ##cpm <- edgeR::cpm(ngs$counts[,samples])
        cpm = 2**ngs$X[,samples,drop=FALSE]
        mean.fc <- sort(rowMeans(cpm),decreasing=TRUE)
        ylab="expression (CPM)"
    }
    if(input$data_type=="logCPM") {
        mean.fc <- sort(rowMeans(ngs$X[,samples,drop=FALSE]),decreasing=TRUE)
        ylab="expression (log2CPM)"
    }

    j <- which(sub(".*:","",names(mean.fc))==gene)
    ##j <- which(ngs$genes$gene_name==gene)
    plot( mean.fc, type="h", lwd=0.4, col="grey80",
         #main="average rank", 
         cex.main=1.2,
         ylab=ylab, xlab="ordered genes", xaxt="n")
    points( j, mean.fc[j], type="h", lwd=2, col="black")
    text( j, mean.fc[j], gene, pos=3, cex=0.9)
})

data_genePlots_waterfallplot_module <- plotModule(
    id="data_genePlots_waterfallplot", func=data_genePlots_waterfallplotFUNC,
    info.text=data_genePlots_waterfallplot_text, pdf.width=10, pdf.height=10, 
    label="d", title="Average rank"
)

output$data_genePlots_waterfallplot     <- data_genePlots_waterfallplot_module$render
output$data_genePlots_waterfallplot_pdf <- data_genePlots_waterfallplot_module$pdf


##----------------------------------------------------------------------
##                     Correlation plot
##----------------------------------------------------------------------

data_genePlots_correlationplotF <- reactive({
    require(RColorBrewer)
    
    ngs <- inputData()
    if(is.null(ngs)) return(NULL)
    
    gene = "KCNN4"
    gene = ngs$genes$gene_name[1]
    if(!is.null(input$search_gene) && input$search_gene!="") gene <- input$search_gene
    samples = colnames(ngs$X)
    if(!is.null(input$data_samplefilter)) {
        samples <- selectSamplesFromSelectedLevels(ngs$Y, input$data_samplefilter)
    }
    nsamples = length(samples)
        
    grp = factor(ngs$Y[samples,"group"])
    klr0 = rep(brewer.pal(8,"Set2"),99)
    klr0 = COLORS
    klr = klr0[factor(ngs$Y[samples,"group"])]

    ## precompute
    pp=rownames(ngs$genes)[1]
    pp <- rownames(ngs$genes)[match(gene,ngs$genes$gene_name)]

    gx = NULL
    ylab = NULL
    if(input$data_type=="counts") {
        gx = ngs$counts[pp,samples]
        ylab="expression (counts)"
    } else if(input$data_type=="CPM") {
        gx = 2**ngs$X[pp,samples]
        ylab="expression (CPM)"
    } else if(input$data_type=="logCPM") {
        gx = ngs$X[pp,samples]
        ylab="expression (log2CPM)"
    }

    par(mar=c(10,3.5,3,1), mgp=c(2.1,0.8,0))

    ## corr always in log.scale and restricted to selected samples subset
    ## should match exactly the rawtable!!
    rho = cor(t(ngs$X[,samples]), ngs$X[pp,samples], use="pairwise")[,1]
    rho[is.na(rho)] <- 0
    jj = head(order(-abs(rho)),30)
    jj <- c( head(order(rho),16), head(order(-rho),16))
    jj <- jj[order(-rho[jj])]
    top.rho = rho[jj]

    gx1 <- sqrt(rowSums(ngs$X[names(top.rho),samples]**2,na.rm=TRUE))
    gx1 <- (gx1 / max(gx1))
    klr1 <- rev(grey.colors(16,start=0.2,end=0.9,gamma=0.25))[1+round(15*gx1) ]
    klr1[which(is.na(klr1))] <- "#DDDDDD"
    names(top.rho) = sub(".*:","",names(top.rho))
    offset = min(top.rho)*0.95
    offset = 0
    res = list(top.rho=top.rho, offset=offset, klr1=klr1)
    return(res)
})

data_genePlots_correlationplotFUNC <- reactive({
    res = data_genePlots_correlationplotF()
    ngs <- inputData()
    if(is.null(ngs) | is.null(res)) return(NULL)
    gene = ngs$genes$gene_name[1]
    if(!is.null(input$search_gene) && input$search_gene!="") gene <- input$search_gene
    barplot(res$top.rho - res$offset, col=res$klr1, ## horiz=TRUE,
            las=3,#main=paste("top correlated genes\nwith",gene),
            main=paste(gene),
            offset = res$offset, ylab="correlation (r)",
            ## names.arg=rep(NA,length(top.rho)),
            cex.names=0.83, cex.main=1.2, border=NA)
    ##text( (1:length(top.rho) - 0.5)*1.2, offset, names(top.rho),
    ##col="black", cex=0.75, srt=90, pos=3, offset=0.4, font=1)
    legend("topright", legend=c("expr high","expr low"),
           fill=c("grey30","grey80"), cex=0.85, y.intersp=0.85)
})

data_genePlots_correlationplot_module <- plotModule(
    id="data_genePlots_correlationplot", func=data_genePlots_correlationplotFUNC,
    info.text=data_genePlots_correlationplot_text, pdf.width=10, pdf.height=10,
    label="c", title="Top correlated genes"
)

output$data_genePlots_correlationplot     <- data_genePlots_correlationplot_module$render
output$data_genePlots_correlationplot_pdf <- data_genePlots_correlationplot_module$pdf


##----------------------------------------------------------------------
##                     Bar/box plot
##---------------------------------------------------------------------- 

data_genePlots_barboxplotFUNC <- reactive({
    require(RColorBrewer)
    
    ngs <- inputData()
    if(is.null(ngs)) return(NULL)
    
    gene = "KCNN4"
    gene = ngs$genes$gene_name[1]
    if(!is.null(input$search_gene) && input$search_gene!="") gene <- input$search_gene
    samples = colnames(ngs$X)
    if(!is.null(input$data_samplefilter)) {
        samples <- selectSamplesFromSelectedLevels(ngs$Y, input$data_samplefilter)
    }
    nsamples = length(samples)
        
    grp = factor(ngs$Y[samples,"group"])
    klr0 = rep(brewer.pal(8,"Set2"),99)
    klr0 = COLORS
    klr = klr0[factor(ngs$Y[samples,"group"])]

    ## precompute
    pp=rownames(ngs$genes)[1]
    pp <- rownames(ngs$genes)[match(gene,ngs$genes$gene_name)]

    gx = NULL
    ylab = NULL
    if(input$data_type=="counts") {
        gx = ngs$counts[pp,samples]
        ylab="expression (counts)"
    } else if(input$data_type=="CPM") {
        gx = 2**ngs$X[pp,samples]
        ylab="expression (CPM)"
    } else if(input$data_type=="logCPM") {
        gx = ngs$X[pp,samples]
        ylab="expression (log2CPM)"
    }

    
    par(mar=c(10,3.5,3,1), mgp=c(2.1,0.8,0))

    bee.cex = ifelse(length(gx)>500,0.1,0.2)
    bee.cex = c(0.3,0.1,0.05)[cut(length(gx),c(0,100,500,99999))]
    ##if(input$data_sampling=="grouped") {
    if(input$data_grouped) {
        ngrp <- length(unique(grp))
        cx1 = ifelse( ngrp < 10, 1, 0.8)
        cx1 = ifelse( ngrp > 20, 0.6, cx1)
        gx.b3plot( gx, grp, las=3, main=gene, ylab=ylab, ## col=klr0[ii], 
                  bar=TRUE, border=NA, ## bee = ifelse(length(gx) < 500,TRUE,FALSE), 
                  bee.cex=bee.cex, ## sig.stars=TRUE, max.stars=5,
                  xlab="", names.cex=cx1, srt=45, cex.main=1.2)
    }  else {
        jj <- 1:length(gx)
        sorting="no"
        ##if(PRO.VERSION) sorting <- input$data_sorting
        if(sorting == "decr")  jj <- order(-gx)
        if(sorting == "inc")  jj <- order(gx)
        tt=""
        barplot(gx[jj], col=klr[jj], las=3, cex.names=0.85,
                ylab=ylab, xlab=tt,
                main=gene, cex.main=1.2, border=NA,
                names.arg=rep(NA,length(gx)) )
        if(length(gx)<100) {
            cx1 = ifelse(length(gx) > 20, 0.85, 1)
            cx1 = ifelse(length(gx) > 40, 0.7, cx1)
            text((1:length(gx)-0.5)*1.2, -0.04*max(gx), names(gx)[jj], 
                 las=3, cex=cx1, pos=2, adj=0, offset=0, srt=60, xpd=TRUE)
        }
    }
})

data_genePlots_barboxplot_module <- plotModule(
    id="data_genePlots_barboxplot", func=data_genePlots_barboxplotFUNC,
    info.text=data_genePlots_barboxplot_text, pdf.width=10, pdf.height=10,
    label="b", title="Abundance/expression barplot"
)

output$data_genePlots_barboxplot     <- data_genePlots_barboxplot_module$render
output$data_genePlots_barboxplot_pdf <- data_genePlots_barboxplot_module$pdf

##----------------------------------------------------------------------
## t-SNE
##----------------------------------------------------------------------

data_genePlots_tsneFUNC <- reactive({
    require(RColorBrewer)
    
    ngs <- inputData()
    if(is.null(ngs)) return(NULL)
    
    gene = "KCNN4"
    gene = ngs$genes$gene_name[1]
    if(!is.null(input$search_gene) && input$search_gene!="") gene <- input$search_gene
    samples = colnames(ngs$X)
    if(!is.null(input$data_samplefilter)) {
        samples <- selectSamplesFromSelectedLevels(ngs$Y, input$data_samplefilter)
    }
    nsamples = length(samples)
        
    grp = factor(ngs$Y[samples,"group"])    
    klr0 = rep(brewer.pal(8,"Set2"),99)
    klr0 = COLORS
    klr = klr0[factor(ngs$Y[samples,"group"])]
    
    ## precompute
    pp=rownames(ngs$genes)[1]
    pp <- rownames(ngs$genes)[match(gene,ngs$genes$gene_name)]
    
    gx = NULL
    ylab = NULL
    if(input$data_type=="counts") {
        gx = ngs$counts[pp,samples]
        ylab="expression (counts)"
    } else if(input$data_type=="CPM") {
        gx = 2**ngs$X[pp,samples]
        ylab="expression (CPM)"
    } else if(input$data_type=="logCPM") {
        gx = ngs$X[pp,samples]
        ylab="expression (log2CPM)"
    }


    #par(mar=c(12,2,2,1), mgp=c(2.1,0.8,0), oma=c(3,0.5,1.5,0.3))
    require(RColorBrewer)
    pos <- ngs$tsne2d[samples,]

    cex1 <- 2*c(1.8,1.2,0.6,0.3)[cut(nrow(pos),breaks=c(-1,40,200,1000,1e10))]    
    klrpal = colorRampPalette(c("blue3", "aliceblue", "grey85", "lavenderblush", "red3"))(16)
    klrpal = colorRampPalette(c("grey80", "grey50", "red3"))(16)

    fc1 <- tanh(0.99 * scale(gx)[,1])
    fc1 <- tanh(0.99 * scale(gx,center=FALSE)[,1])
    ##fc1 <- tanh(0.99 * gx/sd(gx))        
    fc2 <- (fc1 - min(fc1))
    klr1 = klrpal[1 + round(15*fc2/max(abs(fc2)))]
    klr1 = paste0(col2hex(klr1),"88")
    
    #par(mar=c(8,2,2.2,1), mgp=c(1,0.5,0))
    par(mar=c(10,2,3,2), mgp=c(2.1,0.8,0))
    jj2 <- order(abs(fc1))
    plot( pos[jj2,], pch=20, cex=cex1, col=klr1[jj2], fg = gray(0.6), bty = "o",
         xaxt='n', yaxt='n', xlab="tSNE1", ylab="tSNE2")

    grp <- factor(ngs$samples[samples,]$group)
    ngrp <- length(unique(grp))
    if("cell.type" %in% colnames(ngs$samples) && ngrp>20) grp <- ngs$samples[samples,]$cell.type

    cex2 = ifelse(nrow(pos) < 50, 1.5, 1.1)
    cex2 = ifelse(nrow(pos) > 200, 0.8, cex2)
    if(0 && input$pr_labelmode=="legend") {
        legend("bottomright", legend=levels(grp), fill=klrpal,
               cex=cex2, y.intersp=0.8, bg="white")
    } else {
        ##grp.pos <- apply(pos,2,function(x) tapply(x,grp,mean))
        grp.pos <- apply(pos,2,function(x) tapply(x,grp,median))
        if(length(unique(grp))==1) {
            grp.pos <- matrix(grp.pos,ncol=2)
            rownames(grp.pos) <- unique(grp)
        }
        labels = rownames(grp.pos)
        cex3 <- c(1.4,1.2,1,0.8)[cut(length(labels),breaks=c(-1,5,10,20,999))]
        boxes = sapply(nchar(labels),function(n) paste(rep("\u2588",n),collapse=""))
        text( grp.pos, labels=boxes, cex=0.9*cex3, col="#CCCCCC88")
        text( grp.pos, labels=labels, font=2, cex=0.9*cex3, col="black")
        ##text( grp.pos[,], labels=rownames(grp.pos), font=2, cex=cex1**0.5)
    }
    #title("t-SNE clustering", cex.main=1.2, line=0.8)
})

data_genePlots_tsne_module <- plotModule(
    id="data_genePlots_tsne", func=data_genePlots_tsneFUNC,
    info.text=data_genePlots_tsne_text, pdf.width=10, pdf.height=10,
    label="a", title="t-SNE clustering"
)

output$data_genePlots_tsne     <- data_genePlots_tsne_module$render
output$data_genePlots_tsne_pdf <- data_genePlots_tsne_module$pdf


##----------------------------------------------------------------------
##                    Tissue expression plot
##----------------------------------------------------------------------

data_tissueplotFUNC <- reactive({

    ngs <- inputData()
    if(is.null(input$data_type)) return(NULL)
    
    gene <- input$search_gene
    pp <- rownames(ngs$genes)[match(gene,ngs$genes$gene_name)]    
    hgnc.gene = toupper(as.character(ngs$genes[pp,"gene_name"]))
    
    require(RColorBrewer)
    par(mar=c(8,4,2,1), mgp=c(2.2,0.8,0))
    if( hgnc.gene %in% rownames(TISSUE)) {
        tx = TISSUE[hgnc.gene,]
        grp = TISSUE.grp[names(tx)]
        tissue.klr = COLORS[grp]
        ylab="expression (TPM)"
        if(input$data_type=="logCPM") {
            ylab = "expression (log2TPM)"
            tx = log(1 + tx)
        }
        jj <- 1:length(tx)
        sorting="no"
        ##if(PRO.VERSION) sorting=input$data_sorting
        if(sorting=="decr") jj <- order(-tx)
        if(sorting=="inc") jj <- order(tx)
        
        barplot(tx[jj], las=3, main=hgnc.gene, cex.main=1.2,
                col = tissue.klr[jj], border=NA,ylab=ylab, cex.names=0.9, names.arg=rep(NA,length(tx)))
        #title(main="tissue expression", line=-0.3, cex.main=1.1)
        text((1:length(tx)-0.5)*1.2, -0.04*max(tx), names(tx)[jj], las=3,
             cex=0.95, pos=2, adj=0, offset=0, srt=55, xpd=TRUE)
        
    } else { frame()
    }
})

data_tissueplot_module <- plotModule(
    id="data_tissueplot", func=data_tissueplotFUNC,
    info.text=data_tissueplot_text, pdf.width=10, pdf.height=10,
    label="g", title="Tissue expression"
)

output$data_tissueplot     <- data_tissueplot_module$render
output$data_tissueplot_pdf <- data_tissueplot_module$pdf
 
  
##----------------------------------------------------------------------
##                     Correlation plot (stacked)
##----------------------------------------------------------------------

data_corplotF <- reactive({
    require(RColorBrewer)
    ngs <- inputData()
    if(is.null(input$data_type)) return(NULL)
        
    ##samples=colnames(ngs$X);gene="CD4"
    samples <- selectSamplesFromSelectedLevels(ngs$Y, input$data_samplefilter)
    gene <- input$search_gene
    if(is.null(gene)) return(NULL)    
    
    ## corr always in log.scale and restricted to selected samples subset
    zx <- ngs$X
    grp <- ngs$samples$group
    dim(zx)
    if( length(grp) >= 5 && ncol(zx) > 50) {
        zx <- t( apply(ngs$X, 1, function(x) tapply(x,grp,mean)))
    }
    head(zx)
    rownames(zx) <- toupper(sub(".*:","",rownames(zx)))  ## NEED RETHINK!
    
    xref <- list("this_data"=zx, HPA_tissue=as.matrix(TISSUE),
                 ImmProt=as.matrix(IMMPROT))
    gene0 <- toupper(gene)  ## uppercase mouse
    R <- pgx.getGeneCorrelation(gene0, xref=xref)    
    if(is.null(R)) return(NULL)
    
    rho.genes = as.character(ngs$genes$gene_name)
    if("hgnc_symbol" %in% colnames(ngs$genes)) {
        rho.genes = as.character(ngs$genes$hgnc_symbol)
    }
    R <- R[match(rho.genes,rownames(R)),,drop=FALSE]
    rownames(R) <- rho.genes
            
    ## get top correlated genes
    ##jj = head(order(rowSums(R),decreasing=FALSE),35)
    rsum <- rowSums(R,na.rm=TRUE)
    jj = head(order(abs(rsum),decreasing=TRUE),35)
    jj = head(order(-abs(rsum)),30)
    jj <- c( head(order(rsum),20), head(order(-rsum),20))
    jj <- jj[order(-rsum[jj])]
    head(rsum[jj])
    Rtop = R[jj,,drop=FALSE]
    rownames(Rtop) = sub(".*:","",rownames(Rtop))
    offset = min(Rtop, na.rm=TRUE)*0.95
    offset=0
    klr <- grey.colors(ncol(Rtop),start=0.3,end=0.7)
    
    res = list(Rtop=Rtop, offset=offset, klr=klr)
    return(res)
})

data_corplotFUNC <- reactive({

    res <- data_corplotF()
    if(is.null(res)) return(NULL)
    
    par(mar=c(8,4,2,1), mgp=c(2.2,0.8,0))
    barplot( t(res$Rtop) - res$offset, col=res$klr, border=NA, ##horiz=TRUE, 
             las=3, cex.names=0.80, ##names.arg=rep(NA,nrow(R)),
             offset = res$offset, ylab="cumulative correlation (r)",            
             #cex.main=1.2, main="cumulative correlation\nwith other data sets"
             )
    if(!is.null(colnames(res$Rtop))) {
        legend("topright", legend=rev(colnames(res$Rtop)), fill=rev(res$klr),
               cex=0.8, y.intersp=0.8)}
})

data_corplot_module <- plotModule(
    id="data_corplot", func=data_corplotFUNC,
    info.text = data_corplot_text, pdf.width=10, pdf.height=10,
    label="f", title="Cumulative correlation"
)

output$data_corplot     <- data_corplot_module$render
output$data_corplot_pdf <- data_corplot_module$pdf


##----------------------------------------------------------------------
##                     Gene information table
##----------------------------------------------------------------------

output$data_geneInfo  <- renderUI({

    require(KEGG.db)
    require(GO.db)
    require(org.Hs.eg.db)
    
    gene="A1BG-AS1"
    gene = "CD4"
    gene <- input$search_gene
    gene = toupper(sub(".*:","",gene))

    eg = "1017"
    eg = names(which(as.list(org.Hs.egSYMBOL)==gene))
    eg <- mget(gene, env=org.Hs.egSYMBOL2EG, ifnotfound=NA)[[1]]
    if(is.na(eg)) eg <- mget(gene, env=org.Hs.egALIAS2EG, ifnotfound=NA)[[1]]
    eg
    eg = eg[1]
    if(is.null(eg) || length(eg)==0) return(NULL)
    
    output = "(gene info not available)"
    if(length(eg)>0 && !is.na(eg)) {
        ##as.list(org.Hs.egSYMBOL)[[eg]]
        info <- getHSGeneInfo(eg)  ## defined in pgx-functions.R
        if(input$data_geneinfo) {
            info <- c(info, getMyGeneInfo(eg, fields="summary"))  ## defined in pgx-functions.R
        }
        output <- c()
        for(i in 1:length(info)) {
            xx <- paste(info[[i]], collapse=", ")
            output[[i]] <- paste0("<b>",names(info)[i],"</b>: ",xx)
        }
        output <- paste(output, collapse="<p>")
    }    
    ##output <- paste0("<div style='background-color: #dde6f0;'>",output,"</div>")
    HTML(output)
})

data_geneInfo_opt = plotModuleButtons("data_geneInfo_opt", just.info=FALSE, no.download=TRUE, label="e", text=data_geneInfo_text,
                                      options=tagList( tipify( checkboxInput('data_geneinfo','gene summary',FALSE),
                                                       "Provide a summary for the selected gene.", placement="top")), title="Gene info")



##----------------------------------------------------------------------
##                     Interface
##----------------------------------------------------------------------

fillCol( flex = c(1.3,1),
        
        fillRow(flex = c(1,1,1,1),
                
                fillCol(flex = c(NA,1),
                        data_genePlots_tsne_module$button,
                        plotOutput("data_genePlots_tsne", height = "100%", width = "100%")),
                
                fillCol(flex = c(NA,1),
                        data_genePlots_barboxplot_module$button,
                        plotOutput("data_genePlots_barboxplot", height = "100%", width = "100%")),
                
                fillCol(flex = c(NA,1),
                        data_genePlots_correlationplot_module$button,
                        plotOutput("data_genePlots_correlationplot", height = "100%", width = "100%")),
                
                fillCol(flex = c(NA,1),
                        data_genePlots_waterfallplot_module$button,
                        plotOutput("data_genePlots_waterfallplot", height = "100%", width = "100%"))

                ),
        
        fillRow(flex = c(1,1.3,1.6),
                
                fillCol( flex = c(NA,0.01, 1),
                         data_geneInfo_opt,
                         br(),
                         htmlOutput('data_geneInfo', class="gene-info-output")),
                
                fillCol(flex = c(NA,1),
                        data_corplot_module$button,
                        plotOutput("data_corplot", height = "100%", width = "100%")),
                
                fillCol(flex = c(NA,1),
                        data_tissueplot_module$button,
                        plotOutput("data_tissueplot", height = "100%", width = "100%"))
                
                )
        )




```

> **Plots** associated with the selected gene in the `Search gene` section. 
> **(a)** The t-SNE clustering of samples.
> **(b)** An abundance or an expression barplot.
> **(c)** The top correlated genes barplot.
> **(d)** An average rank plot.
> **(e)** The table with further information about the selected gene.
> **(f)** A cumulative correlation barplot with other datasets.
> **(g)** The tissue expression barplot.



### Counts 

```{r echo=FALSE, warning=FALSE, fig.height=5, fig.width=10}

##----------------------------------------------------------------------
##                     Info messsages for Counts
##----------------------------------------------------------------------

counts_tab_barplot_text=paste0('A barplot of the total number of counts (abundance) for each group. The samples (or cells) can be ungrouped by unclicking the ',menu_grouped, ' in the main ',menu_options, '.')
counts_tab_boxplot_text=paste0('A boxplot of the total number of counts (abundance) for each group. The samples (or cells) can be ungrouped by unclicking the ',menu_grouped, ' in the main ',menu_options, '.')
counts_tab_histplot_text=paste0('A histogram of the total number of counts (abundance) for each group. The samples (or cells) can be ungrouped by unclicking the ',menu_grouped, ' in the main ',menu_options, '.')
counts_tab_abundanceplot_text=paste0('A barplot showing the percentage of counts in terms of major gene types such as CD molecules, kinanses or RNA binding motifs for each group. The samples (or cells) can be ungrouped by unclicking the ',menu_grouped, ' in the main ',menu_options, '.')
counts_tab_average_countplot_text=paste0('A barplot showing the average count levels of major gene types such as CD molecules, kinanses or RNA binding motifs for each group. The samples (or cells) can be ungrouped by unclicking the ',menu_grouped, ' in the main ',menu_options, '.')


##----------------------------------------------------------------------
##                     Count information barplot
##----------------------------------------------------------------------

counts_tab_barplotFUNC <- reactive({
    res = counts_tabF()
    if(is.null(res)) return(NULL)
    #par(mar=c(20,6,10,6), mgp=c(2.2,0.8,0))
    par(mar=c(10,4,3,4), mgp=c(2.2,0.8,0))

    subtt0=""
    if(!is.null(res$subtt)) subtt0 <- paste0("\n(",paste(res$subtt,collapse=","),")")
    barplot(res$total.counts/1e6, las=3,
            #main=paste("total counts",subtt0), cex.main=1.7,
            col="grey40", cex.names=res$cx1+0.5, cex.lab=1.5, ylab="counts (million)",
            ylim=c(0,max(res$total.counts)/1e6)*1.1 )
})

counts_tab_barplot_module <- plotModule(
    id="counts_tab_barplot", func=counts_tab_barplotFUNC,
    info.text = counts_tab_barplot_text, pdf.width=13, pdf.height=10, res=45,
    label="a",title='Total counts'
)

output$counts_tab_barplot     <- counts_tab_barplot_module$render
output$counts_tab_barplot_pdf <- counts_tab_barplot_module$pdf


##----------------------------------------------------------------------
##                     Count information boxplot
##----------------------------------------------------------------------

counts_tab_boxplotFUNC <- reactive({
    res = counts_tabF()
    if(is.null(res)) return(NULL)
    #par(mar=c(3,3,3,3), mgp=c(2.4,0.7,0), oma=c(1,1,1,1)*0.2 )   
    par(mar=c(10,4,3,4), mgp=c(2.2,0.8,0))

    boxplot(res$log2counts[res$jj,], col="grey70", 
            #main="counts distribution", cex.main=1.6,
            las=3, cex.names=res$cx1+0.6, cex.lab=1.5, ylab="counts (log2)")
})

counts_tab_boxplot_module <- plotModule(
    id="counts_tab_boxplot", func=counts_tab_boxplotFUNC,
    info.text = counts_tab_boxplot_text, pdf.width=13, pdf.height=10, res=50,
    label="b",title='Counts distribution'
)

output$counts_tab_boxplot     <- counts_tab_boxplot_module$render
output$counts_tab_boxplot_pdf <- counts_tab_boxplot_module$pdf


##----------------------------------------------------------------------
##                     Count information histogram
##----------------------------------------------------------------------

counts_tab_histplotFUNC <- reactive({
    res = counts_tabF()
    if(is.null(res)) return(NULL)
    #par(mfrow=c(2,3), mar=c(9,4,3,1.5), mgp=c(2.4,0.7,0), oma=c(1,1,1,1)*0.2 )  
    par(mar=c(10,4,3,4), mgp=c(2.2,0.8,0))

    n=1000
    gx.hist <- function(gx, n=1000, main="",ylim=NULL) {
        jj <- 1:nrow(gx)
        if(length(jj)>n) jj <- sample(jj,n)
        h0 <- hist(as.vector(c(gx[jj],min(gx),max(gx))),
                   breaks=120, main=main, border=FALSE,
                   col="grey", freq=FALSE, ## ylim=ylim,
                   xlim=c(min(gx),max(gx)),
                   xlab="expression (log2)", 
                   cex.names=res$cx1+0.5, cex.lab=1.5)
        i = 1
        for(i in 1:ncol(gx)) {
            h1 <- hist(gx[jj,i], breaks=h0$breaks,plot=FALSE)
            ##lines( h0$mids, h1$density, col=i+1 )
            lines( h0$mids, h1$density, col="black", lwd=0.5 )
        }
    }
    gx.hist(gx=res$log2counts[res$jj,], n=2000) #, main="histogram")
})

counts_tab_histplot_module <- plotModule(
    id="counts_tab_histplot", func=counts_tab_histplotFUNC,
    info.text = counts_tab_histplot_text, pdf.width=13, pdf.height=10, res=50,
    label="c",title='Counts histogram'
)

output$counts_tab_histplot     <- counts_tab_histplot_module$render
output$counts_tab_histplot_pdf <- counts_tab_histplot_module$pdf


##----------------------------------------------------------------------
##                     Count information abundance of major gene types
##----------------------------------------------------------------------

counts_tab_abundanceplotFUNC <- reactive({
    res = counts_tabF()
    if(is.null(res)) return(NULL)
    #par(mar=c(6,4,0,4), mgp=c(2.2,0.8,0))
    par(mar=c(6,4,0,4))
        
    ## ----------- gene type counts
    ymax = max(colSums(res$prop.counts, na.rm=TRUE))
    barplot(res$prop.counts, las=3, #main="abundance of major gene types", cex.main=1.6,
            cex.names=res$cx1+0.4, cex.lab=1.4, 
            ylim=c(0,ymax)*1.6, ylab="abundance (%)" )
    # legend("topleft", legend=rev(rownames(res$prop.counts)),
    #        fill=rev(grey.colors(nrow(res$prop.counts))),
    #        cex=1, y.intersp=0.4, bty="n")
    leg <- legend("topleft", legend=rev(rownames(res$prop.counts)),
           fill=rev(grey.colors(nrow(res$prop.counts))),
           cex=1, y.intersp=0.4, bty="n", plot = FALSE)
    leftx <- leg$rect$left*0.9
    rightx <- leg$rect$right*0.9
    topy <- leg$rect$top
    bottomy <- leg$rect$bottom
    legend(x = c(leftx, rightx), y = c(topy, bottomy), 
           legend=rev(rownames(res$prop.counts)),
           fill=rev(grey.colors(nrow(res$prop.counts))),
           bty="n", cex=1, y.intersp=0.6)

})

counts_tab_abundanceplot_module <- plotModule(
    id="counts_tab_abundanceplot", func=counts_tab_abundanceplotFUNC,
    info.text = counts_tab_abundanceplot_text, pdf.width=13, pdf.height=10, res=50,
    label="d",title='Abundance of major gene types'
)

output$counts_tab_abundanceplot     <- counts_tab_abundanceplot_module$render
output$counts_tab_abundanceplot_pdf <- counts_tab_abundanceplot_module$pdf


##----------------------------------------------------------------------
##                     Count information average count by gene type
##----------------------------------------------------------------------
counts_tab_average_countplotFUNC <- reactive({
    res = counts_tabF()
    if(is.null(res)) return(NULL)
    #par(mar=c(6,4,0,4), mgp=c(2.2,0.8,0))
    par(mar=c(6,4,0,4))

    ymax = max(colSums(res$avg.counts, na.rm=TRUE))
    barplot(res$avg.counts, las=3, #main="average count by gene type", cex.main=1.6,
            cex.names=res$cx1+0.4, cex.lab=1.4,
            ylim=c(0,ymax)*1.6, ylab="average count")
    leg <- legend("topleft", legend=rev(rownames(res$avg.counts)),
                  fill=rev(grey.colors(nrow(res$avg.counts))),
                  cex=1, y.intersp=0.4, bty="n", plot = FALSE)
    leftx <- leg$rect$left*0.9
    rightx <- leg$rect$right*0.9
    topy <- leg$rect$top
    bottomy <- leg$rect$bottom
    legend(x = c(leftx, rightx), y = c(topy, bottomy),
           legend=rev(rownames(res$avg.counts)),
           fill=rev(grey.colors(nrow(res$avg.counts))),
           cex=1, y.intersp=0.6, bty="n")
})

counts_tab_average_countplot_module <- plotModule(
    id="counts_tab_average_countplot", func=counts_tab_average_countplotFUNC,
    info.text = counts_tab_average_countplot_text, pdf.width=13, pdf.height=10, res=50,
    label="e",title='Average count by gene type'
)

output$counts_tab_average_countplot     <- counts_tab_average_countplot_module$render
output$counts_tab_average_countplot_pdf <- counts_tab_average_countplot_module$pdf

counts_tabF <- reactive({
    ngs = inputData()
    if(is.null(ngs)) return(NULL)
    
    validate(need("counts" %in% names(ngs), "no 'counts' in object."))    
    subtt=NULL

    samples = colnames(ngs$X)
    samples <- selectSamplesFromSelectedLevels(ngs$Y, input$data_samplefilter)
    nsamples = length(samples)
    if("counts" %in% names(ngs)) {
        counts = ngs$counts[,samples,drop=FALSE]
    } else {
        cat("WARNING:: no counts table. estimating from X\n")
        counts = pmax(2**ngs$X - 1,0)
        k = grep("lib.size",colnames(ngs$samples))[1]
        if(length(k)>0) {
            libsize = ngs$samples[colnames(counts),k]
            libsize
            counts = t(t(counts) * libsize)
        }
        ##counts <- round(counts)
    }    
    if(sum(is.na(counts))>0) {
        cat("WARNING:: plot counts: counts has missing values!\n")
    }

    ##if(input$data_sampling=="grouped") {
    gr = ngs$Y[samples,"group"]
    grps = sort(unique(gr))
    if(input$data_grouped && length(grps)>1 ) {
        newx = c()
        for(g in grps) {
            mx = rowMeans(counts[,which(gr==g),drop=FALSE], na.rm=TRUE)
            newx = cbind(newx, mx)
        }
        if(NCOL(newx)==1) newx <- matrix(newx,ncol=1)
        rownames(newx) = rownames(counts)
        colnames(newx) = grps
        counts = newx
    }
    
    if(ncol(counts) > 500) {
        kk <- sample( ncol(counts), 400)
        counts <- counts[,kk,drop=FALSE]
        subtt=c(subtt,"random subset")
    }
    colnames(counts) <- substring(colnames(counts),1,24)
    
    gset <- list()
    gg = ngs$genes[rownames(counts),]$gene_name
    tt = ngs$genes[rownames(counts),]$gene_title
    g1 <- gg[grep("^rpl|^rps",gg,ignore.case=TRUE)]
    g2 <- gg[grep("^mrpl|^mrps",gg,ignore.case=TRUE)]
    g3 <- gg[grep("^MT-",gg,ignore.case=TRUE)]
    g4 <- gg[grep("mitochondr",tt,ignore.case=TRUE)]
    gset[["Ribosomal (RPL/RPS)"]] = g1
    gset[["Mitochondrial ribosomal (MRPL/MRPS)"]] = g2
    gset[["Mitochondrial (MT)"]] = g3
    gset[["Other mitochondrial"]] = setdiff(g4,g3)
    jj <- grep("mitochondr|ribosom",names(FAMILIES),invert=TRUE,ignore.case=TRUE)
    gset.other <- lapply( FAMILIES[jj], function(x) setdiff(x, c(g1,g2,g3,g4)))
    gset <- c(gset, gset.other)
    gset <- gset[grep("<all>",names(gset),invert=TRUE)]
    gset <- gset[sapply(gset,length) > 10]
    
    ## Counts per samples, by category
    total.counts = Matrix::colSums(counts,na.rm=TRUE)
    summed.counts = t(sapply(gset, function(f)
        Matrix::colSums(counts[which(gg %in% f),,drop=FALSE], na.rm=TRUE)))
    avg.counts   = t(sapply(gset, function(f)
        Matrix::colMeans(counts[which(gg %in% f),,drop=FALSE], na.rm=TRUE)))
    prop.counts = 100 * t(t(summed.counts) / total.counts)
    
    head(sort(rowSums(prop.counts,na.rm=TRUE),decreasing=TRUE),10)
    head(sort(rowSums(avg.counts,na.rm=TRUE),decreasing=TRUE),10)
    ##jj <- order(-apply(avg.counts,1,sd,na.rm=TRUE))

    jj <- head(order(-rowSums(prop.counts,na.rm=TRUE)),6)
    ##summed.counts <- summed.counts[jj,,drop=FALSE]
    prop.counts <- prop.counts[jj,,drop=FALSE]    
    jj <- head(order(-rowSums(avg.counts,na.rm=TRUE)),6)
    avg.counts  <- avg.counts[jj,,drop=FALSE]
    sorting="no"
    ##if(PRO.VERSION) sorting <- input$data_sorting
    if(sorting=="decr") {
        total.counts <- sort(total.counts, decreasing=TRUE)
        prop.counts  <- prop.counts[,order(-colMeans(prop.counts))]
        avg.counts   <- avg.counts[,order(-colMeans(avg.counts))]
        counts <- counts[,order(-colMeans(counts))]
    }
    if(sorting=="inc") {
        total.counts <- sort(total.counts, decreasing=FALSE)
        prop.counts  <- prop.counts[,order(colMeans(prop.counts))]
        avg.counts   <- avg.counts[,order(colMeans(avg.counts))]
        counts <- counts[,order(colMeans(counts))]
    }
    
    ss <- names(total.counts)
    prop.counts  <- prop.counts[,ss,drop=FALSE]
    avg.counts   <- avg.counts[,ss,drop=FALSE]
    counts <- counts[,ss,drop=FALSE]
        
    log2counts <- log2(1 + counts)
    ##log2counts[which(log2counts==0)] <- NA
    jj <- sample(nrow(counts),100)
    jj <- sample(nrow(counts),1000)

    ## create the plots
    par(mar=c(3,3,3,3), mgp=c(2.4,0.7,0), oma=c(1,1,1,1)*0.2 )    
    cx1=1
    if(length(total.counts)>12) cx1=0.9
    if(length(total.counts)>30) cx1=0.8
    if(length(total.counts)>50) cx1=0.7
    if(length(total.counts)>80) cx1=0.6

    if(1) {
       names(total.counts) <- substring(names(total.counts),1,30)
       colnames(log2counts) <- substring(colnames(log2counts),1,30)
       colnames(prop.counts) <- substring(colnames(prop.counts),1,30)
       colnames(avg.counts) <- substring(colnames(avg.counts),1,30)
    }

    res = list(total.counts=total.counts,subtt=subtt, cx1=cx1,
               log2counts=log2counts, jj=jj, prop.counts=prop.counts,
               avg.counts=avg.counts)

})


fillCol( flex = c(1,1),

        fillRow(flex = c(1,1,1),

                fillCol(flex = c(NA,1),
                        counts_tab_barplot_module$button,
                        plotOutput("counts_tab_barplot", height = "100%", width = "100%")),
                
                fillCol(flex = c(NA,1),
                        counts_tab_boxplot_module$button,
                        plotOutput("counts_tab_boxplot", height = "100%", width = "100%")),
        
                fillCol(flex = c(NA,1),
                        counts_tab_histplot_module$button,
                        plotOutput("counts_tab_histplot", height = "100%", width = "100%"))
        ),

        fillRow(flex = c(1,1),

                fillCol(flex = c(NA,1),
                        counts_tab_abundanceplot_module$button,
                        plotOutput("counts_tab_abundanceplot", height = "100%", width = "100%")),
                
                fillCol(flex = c(NA,1),
                        counts_tab_average_countplot_module$button,
                        plotOutput("counts_tab_average_countplot", height = "100%", width = "100%"))
                                
                )
        )

```

> Plots associated with the **counts**, abundance or expression levels across the samples. 
> **(a)** A barplot of total counts per sample.
> **(b)** A boxplot of total counts per sample.
> **(c)** A histogram of total counts per sample.
> **(d)** An abundance barplot of major gene types per sample.
> **(e)** An average count barplot by gene type per sample.


### Gene table

```{r echo=FALSE}

dropdown_search_gene='<code>Search gene</code>'
menu_grouped='<code>grouped</code>'
menu_options='<code>Options</code>'

data_rawdataTable_text = paste0('Under the <strong>gene table </strong>, the average expression values of genes across the groups can be read. The samples (or cells) can be ungrouped by unclicking the ',menu_grouped, ' in the main ',menu_options, ' to see the exact expression values per sample (or cell).', 'The genes in the table are ordered by the correlation (<b>rho</b> column) with respect to the gene selected by users from the dropdown menu ',dropdown_search_gene, '. <b>SD</b> column reports the standard deviation of expression across samples (or cells).')


data_rawdataTableFUNC <- reactive({
    ## get current view of raw_counts
    ngs = inputData()
    if(is.null(ngs)) return(NULL)

    pp <- rownames(ngs$X)
    if(input$data_type=="counts") {
        x <- ngs$counts[pp,]
    } else if(input$data_type=="CPM") {
        ##x <- 2**ngs$X
        x <- edgeR::cpm(ngs$counts[pp,])
    } else {
        ## log2CPM
        x <- ngs$X[pp,]
    }
    x0=x
    
    ##------------------ select samples
    x <- x[,]
    samples <- colnames(ngs$X)
    samples <- selectSamplesFromSelectedLevels(ngs$Y, input$data_samplefilter)
    samples <- intersect(colnames(x),samples)
    x <- x[,samples,drop=FALSE]
    
    gene = "CD3E"
    gene = "CCR6"
    gene = input$search_gene
    if(is.null(gene) | gene=="" | is.na(gene)) return(NULL)
    ##xgene = sub(".*:","",rownames(x))

    rho = NULL
    if(1) {
        k=1
        ##k <- which(xgene==gene)
        ##logx <- log2(1 + edgeR::cpm(ngs$counts))
        ##logx <- logx[rownames(x),]
        logx <- ngs$X[rownames(x),]
        xgenes <- ngs$genes[rownames(x),"gene_name"]
        k <- which(xgenes==gene)
        rho = cor( t(logx[,samples]), logx[k,samples], use="pairwise")[,1]
        rho = round(rho[rownames(x)], digits=3)
        sdx = round(apply(logx[,samples],1,sd),digits=3)
    }
    
    ##if(input$data_sampling=="grouped") {
    do.grouped <- input$data_grouped
    if(length(samples)>500) do.grouped <- TRUE
    if(do.grouped) {
        group = ngs$Y[colnames(x),"group"]
        allgroups = sort(unique(group))
        newx = c()
        for(gr in allgroups) {
            mx = rowMeans(x[,which(group==gr),drop=FALSE],na.rm=TRUE)
            newx = cbind(newx, mx)
        }
        rownames(newx) = rownames(x)
        colnames(newx) = paste0("[",allgroups,"]")
        x = newx
    }

    x = round( as.matrix(x), digits=3)
    x95 = quantile(as.vector(x0[which(x0>0)]),probs=0.95)
    x99 = quantile(as.vector(x0[which(x0>0)]),probs=0.99)
    
    if(NCOL(x)==0 || nrow(x)==0) return(NULL)
    ##rownames(x) = sub(".*:","",rownames(x))
    xgenes <- ngs$genes[rownames(x),"gene_name"]
    gene.title <- GENE.TITLE[toupper(xgenes)]
    gene.title <- substring(gene.title,1,50)
    x = data.frame( gene=xgenes, title=gene.title, rho=rho, SD=sdx, as.matrix(x), check.names=FALSE)
    if(!is.null(rho)) {
        x = x[order(-abs(x$rho)),,drop=FALSE]
    } else {
        x = x[order(-x$SD),,drop=FALSE]
    }

    ## put selected gene always on top
    j1 <- which(x$gene==gene)
    jj <- c(j1, setdiff(1:nrow(x),j1))
    x <- x[jj,,drop=FALSE]
    
    DT::datatable( x, rownames=FALSE,
                  class = 'compact cell-border stripe hover',
                  extensions = c('Buttons','Scroller'),
                  selection = list(mode='single', target='row', selected=1),
                  options=list(
                      dom = 'lfrtip', 
                      buttons = list('copy','csv',
                                     list(extend='pdf',
                                          pageSize="A3", orientation='landscape', 
                                          filename="datatable", title="")),
                      ##pageLength = 60,##  lengthMenu = c(20, 30, 40, 60, 100, 250),
                      scroller=TRUE, scrollX = TRUE, scrollY = 800, 
                      deferRender=TRUE
                  )  ## end of options.list 
                  ) %>%
        DT::formatStyle(0, target='row', fontSize='11px', lineHeight='70%') %>%
            DT::formatStyle(colnames(x),
                            background = styleColorBar(c(0,x99), 'lightblue'),
                            backgroundSize = '98% 88%',
                            backgroundRepeat = 'no-repeat',
                            backgroundPosition = 'center')
})

data_rawdataTable_module <- tableModule(
    id = "data_rawdataTable", func = data_rawdataTableFUNC,
    info.text = data_rawdataTable_text,
    title="Gene information table"
)

output$data_rawdataTable <- data_rawdataTable_module$render
output$data_rawdataTable_csv <- data_rawdataTable_module$csv

fillCol(flex = c(NA,0.01, 1),
        data_rawdataTable_module$button,
        br(),
        dataTableOutput("data_rawdataTable", height = "100%", width = "100%")
        )
```

> The gene table contains the average expression values of genes across the groups.


### Sample table

```{r echo=FALSE}
data_sampleTable_text = 'More detailed information about the samples and comparisons are reported under the <b>sample table</b> section.'

data_sampleTableFUNC <- reactive({
    ## get current view of raw_counts
    ngs = inputData()
    if(is.null(ngs)) return(NULL)

    ##if(is.null(input$data_samplefilter)) return(NULL)    
    dt <- NULL
    if(input$data_sampletabletype=="sample info") {
        samples <- selectSamplesFromSelectedLevels(ngs$Y, input$data_samplefilter)
        dt <- ngs$samples[samples,]
    }
    if(input$data_sampletabletype=="contrast info") {
        samples <- selectSamplesFromSelectedLevels(ngs$Y, input$data_samplefilter)
        names(ngs$model.parameters)
        dt <- sign(ngs$model.parameters$exp.matrix[samples,])
        colnames(dt) <- sub("_vs_","\nvs ",colnames(dt))
    }
    
    DT::datatable( dt,
                  class = 'compact cell-border stripe hover',
                  rownames = TRUE,
                  extensions = c('Buttons','Scroller'),
                  selection = list(mode='single', target='row', selected=1),
                  options=list(
                      dom = 'lfrtip', 
                      buttons = list('copy','csv',
                                     list(extend='pdf',
                                          ##pageSize="A3", orientation='landscape', 
                                          filename="sampletable", title="")),
                      ##pageLength = 60, ##  lengthMenu = c(20, 30, 40, 60, 100, 250),
                      scroller=TRUE, scrollX = TRUE, scrollY = 800,
                      deferRender=TRUE
                  )) %>%
        DT::formatStyle(0, target='row', fontSize='12px', lineHeight='70%') 

})

data_sampleTable_opts = tagList(
  tipify( radioButtons('data_sampletabletype','type', inline=TRUE, choices=c("sample info","contrast info")),
          "Select the information table: sample info or contrast info", placement="bottom")
)

data_sampleTable_module <- tableModule(
    id="data_sampleTable", func=data_sampleTableFUNC,
    info.text = data_sampleTable_text,
    options = data_sampleTable_opts,
    title="The sample and contrast information table"
)

output$data_sampleTable     <- data_sampleTable_module$render
output$data_sampleTable_csv <- data_sampleTable_module$csv


fillCol(flex = c(NA,0.01, 1),
        data_sampleTable_module$button,
        br(),
        dataTableOutput("data_sampleTable", height = "100%", width = "100%")
        )

```

> The sample and contrast table contain more detailed information about the samples and comparisons, respectively.

`r if(!PRO.VERSION) {"<!--------------------------------------------- "}`

### Resource info

```{r echo=FALSE, warning=FALSE}

datatable_timingsFUNC <- reactive({
    ngs <- inputData()
    if(is.null(ngs)) return(NULL)
    ##if(is.null(ngs$timings)) return(NULL)
    D <- data.frame()
    if(!is.null(ngs$timings)) {
        D <- round(ngs$timings[,1:3],digits=3)
        D <- apply(D, 2, function(x) tapply(x, rownames(D), sum))
        catg <- gsub("^\\[|\\].*","",rownames(D))
        metd <- gsub("^.*\\]","",rownames(D))
        D <- data.frame(category=catg, method=metd, D)
    }
    DT::datatable( D, rownames=FALSE,
                  options = list(dom='tp', pageLength = 100),
                  class = 'compact cell-border stripe hover' ) %>%
        DT::formatStyle(0, target='row', fontSize='11px', lineHeight='70%')
})

datatable_timings_text = 'More detailed information about the execution times of the methods are reported under the <b>timings</b> table.'
datatable_timings_module <- plotModule(
    id="datatable_timings", func=datatable_timingsFUNC,
    plotlib = "datatable", info.text = datatable_timings_text,
    options = NULL, no.download = TRUE, title='Timings'
)
output$datatable_timings <- datatable_timings_module$render

datatable_objectdimsFUNC <- reactive({
    ngs <- inputData()
    if(is.null(ngs)) return(NULL)

    dims1 <- lapply( ngs, dim)
    lens <- lapply( ngs, length)
    dims2 <- t(sapply( ngs[which(!sapply(dims1,is.null)) ], dim))
    kk <- which(sapply(dims1,is.null))
    dims2 <- rbind(dims2, cbind(lens[kk],0))
    colnames(dims2) = c("nrows","ncols")
    D = data.frame( object=rownames(dims2), dims2, check.names=FALSE)
    DT::datatable( D, rownames=FALSE,
                  options = list(dom='t', pageLength = 50), 
                  class = 'compact cell-border stripe hover') %>%
        DT::formatStyle(0, target='row', fontSize='11px', lineHeight='70%')
})

datatable_objectdims_text = 'More detailed information about the dimensions of objects are reported under the <b>object dimensions</b> table.'
datatable_objectdims_module <- plotModule(
    id="datatable_objectdims", func=datatable_objectdimsFUNC,
    plotlib = "datatable", info.text = datatable_objectdims_text,
    options = NULL, no.download = TRUE, title='Object dimensions'
)
output$datatable_objectdims <- datatable_objectdims_module$render


datatable_objectsizeFUNC <- reactive({
    ngs <- inputData()
    if(is.null(ngs)) return(NULL)
    objsize <- sapply(ngs,object.size)
    objsize <- round( objsize/1e6, digits=2)
    D = data.frame( object=names(ngs), "size.Mb"=objsize, check.names=FALSE)
    DT::datatable( D, rownames=FALSE,
                  options = list(dom='t', pageLength = 50), 
                  class = 'compact cell-border stripe hover') %>%
        DT::formatStyle(0, target='row', fontSize='11px', lineHeight='70%')
})

datatable_objectsize_text = 'More detailed information about the sizes of objects are reported under the <b>object sizes</b> table.'
datatable_objectsize_module <- plotModule(
    id="datatable_objectsize", func=datatable_objectsizeFUNC,
    plotlib = "datatable", info.text = datatable_objectsize_text,
    options = NULL, no.download = TRUE, title='Object sizes'
)
output$datatable_objectsize <- datatable_objectsize_module$render


fillRow(
    flex = c(5,1, 2,1, 1.5, 3), ## width = 600,
    
    fillCol(flex = c(NA,0.01, 1),
            datatable_timings_module$button, br(),
            dataTableOutput("datatable_timings", height = "100%", width = "100%")),br(),

    fillCol(flex = c(NA,0.01, 1),
            datatable_objectdims_module$button, br(),
            dataTableOutput("datatable_objectdims", height = "100%", width = "100%")),br(),

    fillCol(flex = c(NA,0.01, 1),
            datatable_objectsize_module$button, br(),
            dataTableOutput("datatable_objectsize", height = "100%", width = "100%")),br()

)

```

> More detailed information about the execution times of the methods as well as the dimensions and sizes of objects are reported in these tables.

`r if(!PRO.VERSION) {"----------------------------------------->"}`

