```{r}
htmlOutput("current_dataset")
```


Home {data-orientation=rows}
===========================================================================================


Inputs {.sidebar data-width=320}
-------------------------------------------------------------------------------------------

```{r intro-title, results='asis', echo=FALSE}

useShinyjs(rmd=TRUE)
useSweetAlert()
SHOWSPLASH=TRUE
## SHOWSPLASH=FALSE

LOGIN_AUTHENTICATION = "none"
##LOGIN_AUTHENTICATION = "register"
##LOGIN_AUTHENTICATION = "password"
CREDENTIALS <- list("test" = "202cb962ac59075b964b07152d234b70")
CREDENTIALS <- list("test" = "123")


##-----------------------------------------------------------------------------
## Row of social media buttons
##-----------------------------------------------------------------------------
social_buttons <- function() {
    div(
        id="social-buttons",
        tagList(
            ##actionButton("action1", "label", icon=icon("github")),
            tipify( tags$a( href="https://omicsplayground.readthedocs.io", icon("book"), target="_blank"),
                   "Read our online documentation at Read-the-docs", placement="top"),
            tipify( tags$a( href="https://youtube.com", icon("youtube"), target="_blank"),
                   "Watch our tutorials on YouTube", placement="top"),
            tipify( tags$a( href="https://github.com/bigomics/omicsplayground",
                           icon("github"), target="_blank"),
                   "Get the source code or report a bug at GitHub", placement="top"),
            tipify( tags$a( href="https://hub.docker.com/r/bigomics/omicsplayground",
                           icon("docker"), target="_blank"),
                   "Pull our docker from Docker", placement="top"),
            tipify( tags$a( href="https://groups.google.com/d/forum/omicsplayground",
                           icon("users"), target="_blank"),
                   "Get help at our user forum", placement="top")            
        )
    )
}
social_buttons()

##-----------------------------------------------------------------------------
## Show current dataset on each page
##-----------------------------------------------------------------------------
curDataSet <- reactive({
    ngs <- inputData()
    if(is.null(ngs)) return(NULL)
    ##HTML("<b>dataset :</b>",ngs$name,"")
    ##HTML("<h3>",ngs$name,"</h3>")
    dname <- gsub("^.*/|[.]pgx","",ngs$name)
    HTML("<div class='current-data'>",dname,"</div>")
})
output$current_dataset <- renderText({ curDataSet() })


##-----------------------------------------------------------------------------
## User interface
##-----------------------------------------------------------------------------

usermode_infotip = "Select BASIC or ADVANCED user mode. The basic mode should be sufficient for most users. The advanced mode presents extra visualization panels and enables more custom options. The advanced mode also has specialized analysis modules for single-cell RNA-seq, immune cell profiling and signature analysis."

userlevels = c("basic","expert")
## if(DEV.VERSION) userlevels = c("basic","expert","developer")
br();
##require(shinyWidgets)
tipify(radioGroupButtons(
    inputId = "main_usermode",
    label = "User interface:",
    choices = userlevels,
    ## selected = "expert",
    status = "warning",
    checkIcon = list(yes = icon("ok", lib = "glyphicon"))),
    usermode_infotip, placement="bottom", options = list(container="body")
    )

## br()
## tipify(
##     selectInput(
##     "main_workflow","Workflow:",
##     choices = c("Basic expression",
##                 "Expert expression")
##     ),
##     usermode_infotip, placement="bottom", options = list(container="body")
## )


##-----------------------------------------------------------------------------
## Dataset info panel
##-----------------------------------------------------------------------------

cat("<br><br><p>Dataset info:</p>")
wellPanel( htmlOutput("dataset_info") )
output$dataset_info <- renderText({
    inf <- selectedDataSetInfo()
    ## cat("length.inf=",length(inf),"\n")
    inf["conditions"] <- gsub("[,]"," ",inf["conditions"])
    if(length(inf)==0) {
        HTML(paste("<p>Please select a dataset<br>"))
    } else {
        HTML(paste("<p><b>",names(inf),"</b>:", inf,"<br>"))
    }
})


##-----------------------------------------------------------------------------
## UPDATING PGX FILE INFO
##-----------------------------------------------------------------------------

## what files to use, heavy or light objects
PGXINFO <- c()
PGXINFO.FILE <- file.path(PGX.DIR,"datasets-info.csv")

if(file.exists(PGXINFO.FILE)) {
    ## info file exists, check and update
    info = read.csv(PGXINFO.FILE, stringsAsFactors=FALSE, row.names=1)
    pgx.files = dir(PGX.DIR, pattern=".pgx$")
    all(pgx.files %in% info$dataset)
    setdiff(pgx.files, info$dataset)
    if(!all(pgx.files %in% info$dataset)) {
        cat("updating PGX file info...\n")
        info <- pgx.scanInfo(pgx.dir=PGX.DIR, inc.progress=TRUE, pgx=info)
        rownames(info) <- NULL
        Sys.chmod(PGXINFO.FILE, mode="0666")
        write.csv(info, file=PGXINFO.FILE)
    }
    ##info = info[match(pgx.files,info$dataset),] 
    rownames(info) <- NULL
    PGXINFO <- info  
} else {
    ## no info file, scan and create
    cat("scanning PGX file info...\n")
    withProgress(message=">>> scanning available data sets...", value=0, {
        info <- pgx.scanInfo(pgx.dir=PGX.DIR, inc.progress=FALSE)
    })
    ##info.FILE <- "pgx-datasets-info.csv"
    ##Sys.chmod("../shiny-dev", mode="0777")
    Sys.chmod(PGXINFO.FILE, mode="0666")
    write.csv(info, file=PGXINFO.FILE)
    PGXINFO <- info
}

tipify(actionButton("loadbutton",label="Load dataset"),
       "Click to load the selected dataset.", placement="bottom")


##=================================================================================
##========================== MODAL DIALOGS ========================================
##=================================================================================

cartoon_list <- list(
    list(slogan="Visual analytics. See and understand", img="data-graph-wisdom.jpg"),
    list(slogan="Fasten your seat belts. Accelerated discovery", img="cartoon-speedup.jpg"),
    list(slogan="Analytics anywhere. Anytime.", img="cartoon-cloudservice.jpg"),
    ## list(slogan="Do-it-yourself. Yes you can.", img="bigomics-rockstar3.jpg"),
    list(slogan="Analyze with confidence. Be a rockstar", img="bigomics-rockstar3.jpg"),
    list(slogan="Fast track your Bioinformatics", img="selfservice-checkout2.png"),
    list(slogan="Your analysis doesn't take coffee breaks", img="gone-for-coffee.png"),
    list(slogan="Big Friendly Omics", img="big-friendly-omics1.jpg"),
    list(slogan="Big Data meets Biology", img="bigdata-meets.png")
)

randomCartoon <- function() {
    ## invalidateLater(12000)
    cartoon <- sample(cartoon_list,1)[[1]]
    cartoon$img2 = paste0("cartoons/",cartoon$img)
    cartoon$img  = paste0("www/cartoons/",cartoon$img)
    cartoon
}

startup_count=0
VERSION='0.99 (always almost)'

require(shinyparticles)
particlesjs.conf <- rjson::fromJSON(file="resources/particlesjs-config.json")

showStartupModal <- function(once=FALSE) {    

    if(length(input$loadbutton)==0) return(NULL)  ## UI not ready
    if(once && startup_count>0) return(NULL)

    dbg("showStartupModal: showing!\n")    
    showModal(modalDialog(
        id="modal-splash",
        div(
            id="particles-target",
            img(src = base64enc::dataURI(file="www/splash/bigomics-splash2.png"),
                width="100%", height="auto%", style = "position:absolute;"),
            style="height: 500px; width: 100%; background-color: #2c81e2;"
            ##height="500px", style="margin-left: -14px; margin-top: -4px;"
        ),
        footer = tagList(
            actionButton("action1","Read-the-docs", icon=icon("book"),
                         onclick="window.open('https://omicsplayground.readthedocs.io','_blank')"),
            actionButton("action2","Watch the tutorial", icon=icon("youtube"),
                         onclick="window.open('https://youtube.com','_blank')"),
            actionButton("action3","Get the source", icon=icon("github"),
                         onclick="window.open('https://github.com/bigomics/omicsplayground','_blank')"),
            actionButton("action4","Docker image", icon=icon("docker"),
                         onclick="window.open('https://hub.docker.com/r/bigomics/omicsplayground','_blank')"),
            actionButton("action5","User forum", icon=icon("users"),
                         onclick="window.open('https://groups.google.com/d/forum/omicsplayground','_blank')"),
            ## actionButton("action_beer","Buy us a beer!", icon=icon("beer")),
            ## actionButton("action5","Donate pizza", icon=icon("pizza-slice")),
            ## modalButton("Let's start!")
            actionButton("action_play", "Let's play!")
        ),
        shinyparticles::particles(config=particlesjs.conf, target_id = "particles-target", timeout = 1000),
        size="l", easyClose=FALSE, fade=FALSE))
    ##startup_count <<- startup_count + 1    
    dbg("showStartupModal done!\n")
}

observeEvent( input$action_beer, {
    dbg("buy beer button action\n")
    startup_count <<- startup_count + 1    
    USER$logged <- TRUE
    USER$name   <- "beer buddy"
    removeModal()
    ##alert("Wow. Thanks buddy!")
    sendSweetAlert(
        session=session, title="Wow. Thanks buddy!",
        text = "Free entrance for you!", type = "info")
    ##Sys.sleep(4);removeModal()
})


observeEvent( input$action_play, {

    dbg("action_play:: play button action\n")
    cat("action_play:: LOGIN_AUTHENTICATION=",LOGIN_AUTHENTICATION,"\n")

    startup_count <<- startup_count + 1    
    if(LOGIN_AUTHENTICATION!="none") {
        showLogin()  ## $
    } else {
        removeModal()
    }
})

selectedDataSetInfo <- reactive({
    sel <- input$pgxtable_rows_selected
    if(is.null(sel) || length(sel)==0) return(NULL)
    df <- getPGXTable()
    unlist(lapply(df[sel,],as.character))
})

selectedDataSet <- reactive({
    sel <- input$pgxtable_rows_selected
    if(is.null(sel) || length(sel)==0) return(NULL)
    df <- getPGXTable()
    as.character(df$dataset[sel])
})

showLoadingModal <- function(msg="Loading data...") {
    toon <- randomCartoon()
    showModal(modalDialog(
        ##title = HTML("<center><h4>Omics Playground</h4></center>"),
        ##HTML("<center><h2>",pgx.randomSlogan(),"</h2><h4>with Omics Playground</h4></center>"),
        title = HTML("<center><h2>",toon$slogan,"</h2><h4>with Omics Playground</h4></center>"),
        fillRow(flex=c(1,NA,1), br(),
                ##imageOutput("loading_image", width="auto", height="250px"),
                img(src = base64enc::dataURI(file=toon$img), width="auto", height="250px"),
                br()),
        footer = HTML("<center><p>",msg,"  &nbsp; Please wait</p></center>"),
        size="m", easyClose=TRUE, fade=TRUE))
    ## Sys.sleep(5)
    dbg("showLoadingModal done!\n")
}

currentSection <- reactive({
    cdata <- session$clientData
    sub("section-","",cdata[["url_hash"]])
})


##=================================================================================
##==================== USER AUTHENTICATION ========================================
##=================================================================================


USER <- reactiveValues( logged = FALSE, name="anonymous")

observeEvent( input$logout, {
    ##updateTextInput(session, ".username", value=NULL)
    reset("login_username")
    reset("login_password")
    USER$logged <- FALSE
})

if(LOGIN_AUTHENTICATION=="password") {

    showLogin <- function() {
        showModal( modalDialog(
            id = "login",
            title = "Login to Omics Playground",
            tagList( 
                textInput("login_username", "Username:"),
                passwordInput("login_password", "Password:"),
                div( actionButton("login_btn", "Login"),
                    style="text-align: center;")
            ),
            footer = textOutput("login_warning"),
            size = "s"
        ))
    }


} else if(LOGIN_AUTHENTICATION=="register") {    

    showLogin <- function() {
        showModal( modalDialog(
            id = "login",
            title = "Login to Omics Playground",
            tagList( 
                p("Please",actionLink("register_link","register"),"to use the Omics Playground. If you have already registered please enter your registration email address below."),
                textInput("login_username", "E-mail:"),
                div( 
                    actionButton("login_btn", "Login"),
                    style="text-align: center;")
            ),
            footer = div(textOutput("login_warning"),style="color: red;"),
            size = "s"
        ))
    }
    
    observeEvent( input$register_link, {
        ##install.packages("countrycode")
        require(countrycode)
        all.countries = countrycode::codelist[,"country.name.en"]
        
        showModal( modalDialog(
            id = "register",
        title = "Omics Playground registration",
        tagList( 
            p("Fill in the form below. Registration data is used to measure usage only. This helps us track and better serve our user community."),
            ## textInput("register_name", "Name:"),
            textInput("register_email", "E-mail:"),
            ## textInput("register_organization", "Organization:"),
            selectInput("register_country", "Country:", choices=c("",all.countries)),
            div( ##actionButton("register_btn_skip", "Skip"),
                actionButton("register_btn", "Register"),
                style="text-align: center;")
        ),
        footer = div(textOutput("register_warning"),style="color: red;"),
        size = "s"
        ))
    })           

    observeEvent( input$register_btn, {
        register.OK = ( input$register_email != "" &&
                        grepl("@",input$register_email) &&  ## valid email
                        ## input$register_name != "" &&
                        input$register_country != "" )
        if(register.OK) {
            rdata <- paste("Tom Cruise","ACME Inc.","USA",date(),sep=",")
            rdata <- paste(input$register_email,
                           input$register_name,
                           input$register_organization,
                           input$register_country,
                           date(), sep=",")
            write( rdata, file="logs/registered.csv", append=TRUE )

            USER$name   <- input$register_email
            USER$logged <- TRUE  ## global reactive
            show("register_warning")
            output$register_warning = renderText(paste("Welcome",input$register_email,"!"))
            delay(3000, hide("register_warning", anim = TRUE, animType = "fade"))
            removeModal()
            
        } else {
            ##show("register_warning")
            output$register_warning = renderText("Invalid email or country")
            ##delay(2000, hide("register_warning", anim = TRUE, animType = "fade"))
            delay(2000, {output$register_warning <- renderText("")})
        }
    })
} else {
    showLogin <- function() {}
}

output$login_warning = renderText("")

observeEvent( input$login_btn, {           

    cat("LOGIN_AUTHENTICATION=",LOGIN_AUTHENTICATION,"\n")
    cat("username=",input$login_username,"\n")
    cat("password=",input$login_password,"\n")
    cat("Logged=",USER$logged,"\n")

    login.OK = FALSE
    if(LOGIN_AUTHENTICATION=="register") {
        ##if( is.null(input$login_username) || is.null(input$login_password)) return(NULL)
        ##if( input$login_username=="" || input$login_password=="") return(NULL)    
        ##if( is.null(input$login_name) || input$login_name == "") return(NULL)
        if( is.null(input$login_username) || input$login_username=="") return(NULL)
        registered <- read.csv(file="logs/registered.csv",header=FALSE,stringsAsFactors=FALSE)
        registered.users <- unique(registered[,1]) ## emails

        cat("registered.users=",registered.users,"\n")
        registered.users <- c(registered.users,"demo@bigomics.ch")
        login.OK = (input$login_username %in% registered.users)
    }
    if(LOGIN_AUTHENTICATION=="password") {
        if( is.null(input$login_username) || is.null(input$login_password)) return(NULL)
        if( input$login_username=="" || input$login_password=="") return(NULL)    
        login.OK = isTRUE(CREDENTIALS[[input$login_username]]==input$login_password)
    }

    if (login.OK) {
        output$login_warning = renderText("")
        removeModal()

        USER$name   <- input$login_username
        USER$logged <- TRUE
        
        ## Here you can perform some user-specific functions, or site news
        if(0) {
            showModal(modalDialog(
                paste("Welcome",USER$name,"!"),
                footer=NULL, size="m"))
            Sys.sleep(3)
        }
        removeModal()
        
    } else {
        ##show("login_warning")
        if(LOGIN_AUTHENTICATION=="password") {
            output$login_warning = renderText("Invalid username or password")
        }
        if(LOGIN_AUTHENTICATION=="register") {
            output$login_warning = renderText("Email address not recognized")
        }
        ##delay(2000, hide("login_warning", anim = TRUE, animType = "fade"))
        delay(2000, {output$login_warning <- renderText("")})
        USER$logged <- FALSE
    }
    ##hide("login_warning")
})
   
##=================================================================================
##======================== USER LEVEL =============================================
##=================================================================================

##-----------------------------------------------------------------------------
## Select UI user level
##-----------------------------------------------------------------------------

BASIC_SECTIONS <- list(
    c('#navbar','#section-dataview'),
    c('#navbar','#section-clustering'),
    c('#navbar','#section-expression'),
    c('#navbar','#section-enrichment'),
    c('#navbar','#section-intersection'),
    c('#navbar','#section-functional')
)

EXPERT_SECTIONS <- list(
    ## c('#navbar','#section-test'),
    c('#navbar','#section-signature'),
    c('#navbar','#section-biomarker'),
    c('#navbar','#section-scprofiling'),
    c('#section-clustering','#section-feature-ranking'),
    c('#section-expression','#section-fdr-table'),
    c('#section-expression','#section-volcano-methods'),
    c('#section-enrichment','#section-fdr-table-1'),
    c('#section-enrichment','#section-volcano-methods-1'),
    c('#section-intersection','#section-connectivity-map'),
    c('#section-intersection','#section-meta-volcano'),
    c('#section-functional','#section-drug-cmap'),
    c('#section-functional','#section-wordcloud')
)

USERMODE = factor("basic", levels=c("intruder","basic","expert","developer","jedi","god"))

observeEvent(input$main_usermode,
{
    if(input$main_usermode == "basic") {
        dbg("observeEvent::main_usermode : switching to BASIC mode")
        lapply(c(EXPERT_SECTIONS), function(sec) {
            sel = paste0(sec[1]," li a[href='",sec[2],"']")
            dbg("observeEvent::main_usermode : hiding ",sel)
            shinyjs::hide(selector = sel)
        })
        ##shinyjs::html("navbar-brand","Omics Playground (basic)")
        USERMODE <<- "basic"
    }
    
    if(input$main_usermode >= "expert") {
        dbg("observeEvent::main_usermode : switchhing to EXPERT mode")
        lapply(EXPERT_SECTIONS, function(sec) {
            sel = paste0(sec[1]," li a[href='",sec[2],"']")
            dbg("observeEvent::main_usermode : showing ",sel)
            shinyjs::show(selector = sel)
        })
        USERMODE <<- "expert"        
    }
    
    ##updateNavbarPage(session, "navbar", selected=1)
})

##================================================================================
##====================== INPUT DATA REACTIVE OBJECT ==============================
##================================================================================

##inputData <- eventReactive( input$loadbutton, {
inputData <- reactive({
    ##-----------------------------------------------------------------
    ## This is the main loader function that loads the ngs object.
    ##-----------------------------------------------------------------
    dbg("inputData:: reacted\n")
    dbg("inputData:: LOGIN_AUTHENTICATION=",LOGIN_AUTHENTICATION,"\n")
    dbg("inputData:: USER$Logged=",USER$logged,"\n")
    
    ## authenicate user
    ##if(!USER$logged) showLogin()
    if(LOGIN_AUTHENTICATION!="none" && !USER$logged) return(NULL)
    
    ## Sense button press
    input$loadbutton   
    btn <- isolate(input$loadbutton)
    dbg("inputData:: input$loadbutton=",btn,"\n")
    dbg("inputData:: length(btn)=",length(btn),"\n")

    pgx = NULL
    pgx = isolate(selectedDataSet())
    dbg("inputData:: 1: pgx.selected=",pgx)
    if(!is.null(btn) && btn!=0 && !is.null(pgx)) {
        dbg("inputData:: showing LOADING")
        ## show loading pop-up
        showLoadingModal()
    }
    if(is.null(pgx) || pgx=="" || length(pgx)==0) {
        ## Set to default data set if table is not ready yet.
        pgx <- PGXINFO$dataset[1]
        dbg("<inputData:reactive>: setting to default data set = ",pgx,"\n")
    }
    
    pgx1 = file.path(PGX.DIR,pgx)
    if(file.exists(pgx1)) {
        dbg("inputData:: loading",pgx1,"\n")
        ##withProgress(message='loading...', value=0.8,
        load(pgx1,verbose=0)
        dbg("inputData:: finished loading\n")
    } else {
        cat("inputData:: ERROR file not found : ",pgx1,"\n")
        removeModal()
        return(NULL)
    }
    
    ##----------------- update input
    ngs <- pgx.initialize(ngs)

    if(is.null(ngs$name)) ngs$name <- sub("[.]pgx$","",pgx)

    ##----------------- remove modal??
    if(startup_count>0) {
        dbg("inputData:: removing modaldialog\n")
        Sys.sleep(4)
        removeModal()
    }
    dbg("inputData:: ready!\n")
    return(ngs)

})
##}, ignoreNULL=FALSE )
##}, ignoreNULL=TRUE )


##=================================================================================
##=================================================================================
##=================================================================================

```

Row {data-height=120}
-----------------------------------------------------------------------

```{r}
require(shinydashboard)
## useShinydashboard()
fillRow(
    height=115,
    ##renderValueBox( div(valueBox(value=200, "subtitle", color="light-blue"), class="value-box")),
    div( uiOutput("valuebox1"), style="height: 110px; background-color: rgba(39, 128, 227, 0.7); margin-right: 4px;"),
    div( uiOutput("valuebox2"), style="height: 110px; background-color: rgba(39, 128, 227, 0.7); margin-left: 4px; margin-right: 4px;"),
    div( uiOutput("valuebox3"), style="height: 110px; background-color: rgba(39, 128, 227, 0.7); margin-left: 4px; margin-right: 10px;")
)

vbox <- function(value, label) {
    box(h1(value, style="font-weight: 800; color: white; padding-top: 16px; margin-top: 0px; margin-left: 20px;"),
        h5(label, style="color: white; margin-left: 20px;"), width="100%")    
}

output$valuebox1 <- renderUI({
    pgx <- getPGXTable()
    req(pgx)
    ndatasets <- nrow(pgx)
    vbox( ndatasets, "data sets")     
})
output$valuebox2 <- renderUI({
    pgx <- getPGXTable()
    req(pgx)
    nsamples <- sum(pgx$nsamples)
    vbox( nsamples, "number of samples")     
})
output$valuebox3 <- renderUI({
    pgx <- getPGXTable()
    req(pgx)
    nvalues <- sum(pgx$nsamples * (pgx$ngenes + pgx$nsets))
    nvalues1 <- format(nvalues, nsmall=, big.mark=" ")  # 1,000.6
    vbox( nvalues1, "data points") 
})

br();br();

```


Row {.tabset data-height=850}
-----------------------------------------------------------------------

### Data sets

```{r}

##split=" ";n=5
andothers <- function(s, split=" ", n=8) {
    if(is.na(s)) return("")
    s1 <- strsplit(s, split=split)[[1]]
    if(length(s1)<=n) return(s)
    n2 <- setdiff(length(s1),n)
    paste(paste(head(s1,n), collapse=" "),"(+",n2,"others)")
}


## Some 'global' reactive variables used in this file
uploaded_files <- reactiveValues()
uploadReady    <- reactiveVal(0)
##shinyjs::disable("upload_compute")

getPGXTable <- reactive({
    ## get table of data sets
    ##
    ##
    
    if(is.null(PGXINFO)) return(NULL)    
    dbg("<getPGXTable> reacted")
    
    has.upload = uploadReady() ## reactive
    ##has.upload <- ("ngs" %in% uploaded_files) ## reactive
    dbg("<getPGXTable> has.upload=",has.upload)
    dbg("<getPGXTable> 1")
    
    df <- PGXINFO
    pgx.files = dir(PGX.DIR, pattern=".pgx$")
    sel <- sub("[.]pgx$","",df$dataset) %in% sub("[.]pgx$","",pgx.files)
    df <- df[sel,]
    
    ##kk = unique(c("dataset","datatype","organism","description",colnames(df)))
    kk = unique(c("dataset","datatype","organism","description","nsamples",
                  "ngenes","nsets","conditions","date"))
    kk = intersect(kk,colnames(df))
    df = df[,kk]

    dbg("<getPGXTable> 2")
    
    df = df[order(df$dataset),]   ## sort alphabetically...
    rownames(df) <- NULL

    df
})


pgxTable.RENDER <- reactive({

    dbg("<pgxTable.RENDER> reacted")
    df <- getPGXTable()
    req(df)
    
    if(SHOWSPLASH) showStartupModal(once=TRUE)

    has.upload = uploadReady() ## reactive
    
    dbg("<pgxTable.RENDER> head(df$dataset)=",head(df$dataset))
    
    df$conditions  <- gsub("[,]"," ",df$conditions)
    df$conditions  <- sapply(as.character(df$conditions), andothers, split=" ", n=5)
    df$description <- shortstring(as.character(df$description),200)
    df$nsets <- NULL
    df$date  <- NULL
    
    DT::datatable( df,
                  class = 'compact cell-border stripe hover',
                  rownames=TRUE,
                  extensions = c('Scroller'),
                  ##selection = list(mode='single', target='row'),
                  selection = list(mode='single', target='row', selected=1 ),
                  ##selection = list(mode='none'),
                  ##filter = "top",
                  options=list(
                      ##dom = 'Blfrtip', 
                      pageLength = 100, ##  lengthMenu = c(20, 30, 40, 60, 100, 250),
                      scrollX = FALSE, scrollY =580, ## scroller=TRUE,
                      deferRender=TRUE
                  )  ## end of options.list 
                  )  %>%
        DT::formatStyle(0, target='row', fontSize='12px', lineHeight='95%')

})

pgxDatasetOverview.RENDER <- reactive({
    
    df <- getPGXTable()
    req(df)
    sel = input$pgxtable_rows_selected    
    sel.dataset = df$dataset[sel]
    par(mfrow=c(1,3))
    plot(sin)
    plot(cos)
    plot(exp)
    
})


pgxtable_text = "This table contains a general information about all available datasets within the platform. For each dataset, it reports a brief description as well as the total number of samples, genes, gene sets (or pathways), corresponding phenotypes and the collection date."

pgxtable_module <- tableModule(
    id = "pgxtable", func = pgxTable.RENDER,
    info.text = pgxtable_text,
    title="Datasets"
    )
output <- attachModule(output, pgxtable_module)
outputOptions(output, "pgxtable", suspendWhenHidden=FALSE) ## important!

fillCol(
    ##flex=c(1,3),
    ##moduleWidget(pgxplots_module, outputFunc="plotOutput")
    moduleWidget(pgxtable_module, outputFunc="dataTableOutput")
)

```


### Upload data

```{r data-upload, echo=FALSE}
upload_filetypes = c("text/csv","text/comma-separated-values,text/plain",".csv")

output$downloadExampleData <- downloadHandler(
    filename = "exampledata.zip",
    content = function(file) {
        zip = paste0(PGX.DIR,"exampledata.zip")
        file.copy(zip,file)
    }
)

upload_info = "<h4>User file upload</h4><p>Please prepare the data files in CSV format as listed below. It is important to name the files exactly as shown. The file format must be comma-separated-values (CSV) text. Be sure the dimensions, rownames and column names match for all files. You can download a zip file with example files here: EXAMPLEZIP. After that, in sidebar on the left, provide a unique name and description for your dataset. Finally, hit the compute button. The computations may take 10 to 30 minutes depending on the size of your dataset.<br><br>"
DLlink = downloadLink("downloadExampleData","exampledata.zip")
upload_info = sub("EXAMPLEZIP",DLlink,upload_info)

upload_info2 =
"<br><h4>Uploaded datasets</h4><p>Below are your uploaded datasets. As a free user, you can only have a maximum of one private dataset. If you want to analyze a new dataset, you must either delete your old dataset or make the dataset public.<br><br>"

fillRow(
    flex=c(1,0.05,3),
    wellPanel(
        fileInput("upload_files", "Choose files",
                  multiple = TRUE, accept = upload_filetypes),
        textInput("upload_name","Name of dataset:"),
        selectInput("upload_datatype","Datatype:",
                    choices=c("RNA-seq","proteomics","scRNA-seq","multi-omics")),
        textAreaInput("upload_description", "Description:", value = NULL,
                      rows=5, placeholder="Describe your data set (minimum 100 characters)"),
        ##conditionalPanel(
        ##"output.allFilesOK",
        actionButton("upload_compute","Compute!",icon=icon("running"))
        ##)
    ),br(),
    fillCol(
        ##flex = c(NA,NA,NA,NA,1),
        flex = c(NA,NA,1),
        p(HTML(upload_info)),
        tableOutput("upload_status"),
        ##p(HTML(upload_info2)),
        ##DT::dataTableOutput("upload_datasetsDT"),
        ##tableOutput("upload_datasets")
        br()
    )
)

output$allFilesOK <- reactive({    
    files.needed = c("counts.csv","samples.csv","genes.csv","contrasts.csv")
    all.there <- all(files.needed %in% names(uploaded_files))    
    df <- uploadStatusTable()
    all.ok = all( df$status == "OK")
    filled = ( input$upload_name!="" && nchar(input$upload_description)>=100)
    ##all.ok    <- all(sapply(uploaded_files, function(x) x$status=="OK"))
    ok <- (all.there && all.ok && filled)
    if(ok) shinyjs::enable("upload_compute")
    if(!ok) shinyjs::disable("upload_compute")
    ok
})
outputOptions(output, "allFilesOK", suspendWhenHidden = FALSE) ## important!

observeEvent( input$upload_compute, {

    if(0) {
        
        files.needed = c("counts.csv","samples.csv","genes.csv","contrasts.csv")
        files.needed1 = paste0("../exampledata/",files.needed)
        uploaded_files = lapply(files.needed1, read.csv, row.names=1,
                                check.names=FALSE, stringsAsFactors=FALSE)
        names(uploaded_files) <- files.needed
        uploaded_files[["counts.csv"]] <- as.matrix(uploaded_files[["counts.csv"]])
        uploaded_files[["contrasts.csv"]] <- as.matrix(uploaded_files[["contrasts.csv"]])
        for(i in 1:length(uploaded_files)) {
            colnames(uploaded_files[[i]]) <- gsub("[\n\t ]","_",colnames(uploaded_files[[i]]))
            rownames(uploaded_files[[i]]) <- gsub("[\n\t ]","_",rownames(uploaded_files[[i]]))
        }

    }

    ## check names and other things
    dbg("upload_compute :: checking filename")

    pgx.files = sub("[.]pgx$","",dir(PGX.DIR, pattern=".pgx$"))
    if(input$upload_name %in% pgx.files) {
        sendSweetAlert(
            session = session,
            title = "File exists!",
            text = "Please use a different name.",
            type = "warning"
        )
        return(NULL)
    }

    ## --------------------- OK start ---------------------------
    dbg("upload_compute :: showing coffee modal")
    showLoadingModal("Calculating. It's a good time to get a coffee now...")

    counts    <- as.matrix(uploaded_files[["counts.csv"]])
    samples   <- data.frame(uploaded_files[["samples.csv"]],stringsAsFactors=FALSE)
    genes     <- data.frame(uploaded_files[["genes.csv"]],stringsAsFactors=FALSE)
    contrasts <- as.matrix(uploaded_files[["contrasts.csv"]])

    if(ncol(counts) > 1000) {
        gx.methods   = c("ttest.welch","trend.limma","edger.qlf")
        gset.methods = c("fisher","gsva","fgsea")
    } else {     
        gx.methods   = c("trend.limma","edger.qlf","deseq2.wald")
        gset.methods = c("fisher","gsva","fgsea")
        ##gset.methods = c("fisher","gsva","spearman")
    }

    extra.methods = c("meta.go","deconv","infer","drugs")
    ##extra.methods = c("meta.go","infer")

    if(0) {
        ## just for debugging...
        ##
        dbg("upload_compute :: dummy computation...")
        load("../data/geiger2016-arginine.pgx")
    } else {

        dbg("upload_compute :: ***** real computation *****")
        ##----------------------------------------------------------------------
        ## Limit the data dimensions
        ##----------------------------------------------------------------------
        if(1) {
            contrasts <- contrasts[,1:min(3,ncol(contrasts))]   ## maximum nr of contrasts...
            counts    <- counts[,1:min(100,ncol(counts))]       ## maximum nr of samples
            samples   <- samples[colnames(counts),,drop=FALSE]
            ##contrasts <- contrasts[colnames(counts),,drop=FALSE]
            ii <- head(order(-apply(log(10+counts), 1, sd, na.rm=TRUE)),2000) ## maximum nr of genes
            ##ii <- head(order(-apply(log(10+counts), 1, sd, na.rm=TRUE)),8000) ## maximum nr of genes
            counts <- counts[ii,,drop=FALSE]
            genes  <- genes[rownames(counts),,drop=FALSE]
        }
        
        dbg("upload_compute :: dim(counts)=",dim(counts))
        dbg("upload_compute :: class(counts)=",class(counts))
        dbg("upload_compute :: sum.is.na(counts)=",sum(is.na(counts)))

        dbg("upload_compute :: dim(samples)=",dim(samples))
        dbg("upload_compute :: dim(genes)=",dim(genes))
        dbg("upload_compute :: dim(contrasts)=",dim(contrasts))
        dbg("upload_compute :: class(contrasts)=",class(contrasts))
        
        start_time <- Sys.time()
        ## Create a Progress object
        progress <- shiny::Progress$new()
        on.exit(progress$close())    
        progress$set(message = "Processing data set", value = 0)

        ngs <- pgx.upload(counts, samples, genes, contrasts,
                          progress=progress,
                          gx.methods = gx.methods,
                          gset.methods = gset.methods,
                          extra.methods = extra.methods
                          )

        end_time <- Sys.time()
        delta_time  = end_time - start_time
        delta_time
        dbg("upload_compute :: total processing time of",delta_time,"secs")
        
        names(ngs)
        head(ngs$samples)
    }

    ngs.name = "testdata"
    ngs.name = gsub("[ ]","-",input$upload_name)
    pgx.name = paste0(ngs.name,".pgx")
    ngs$name = ngs.name
    ngs$datatype = input$upload_datatype
    ngs$description = input$upload_description
    ngs$date = date()

    ## -------------------- save PGX file/object
    ## ngs <- ngs.initizalize(ngs)
    ##uploaded_files[["ngs"]] <- ngs
    fn = paste0(PGX.DIR,ngs.name,".pgx")
    fn
    dbg("upload_compute :: saving PGX to",fn)
    save(ngs, file=fn)  ## would clash with multiple users...

    ## -------------------- update PGXINFO table
    ##pgx.dir=PGX.DIR;inc.progress=FALSE;pgx=PGXINFO
    old.info = PGXINFO[which(!PGXINFO$dataset %in% c(ngs.name,pgx.name)),,drop=FALSE]
    new.info <- pgx.scanInfo(pgx.dir=PGX.DIR, inc.progress=FALSE, pgx=old.info)
    PGXINFO <<- new.info
    Sys.chmod(PGXINFO.FILE, mode="0666")
    write.csv(PGXINFO, file=PGXINFO.FILE)
           
    removeModal()

    uploadReady(uploadReady() + 1)  ## notify reactive variable
    ##selectRows(proxy = dataTableProxy("pgxtable"), selected=1)
    ##shinyjs::click("loadbutton")    

    ## showModal(modalDialog(p("Ready! Now go back and select your uploaded data set.")))
    sendSweetAlert(
        session = session,
        title = "Ready !!",
        text = "Now go to the previous tab and select your uploaded data set.",
        type = "success"
    )
    
})


uploadStatusTable <- reactive({
    
    cat("<uploaded_files> name=",input$upload_files$name,"\n")
    cat("<uploaded_files> datapath=",input$upload_files$datapath,"\n")

    ## read uploaded files
    ff = lapply(input$upload_files$datapath, read.csv, row.names=1,
                check.names=FALSE, stringsAsFactors=FALSE )
    names(ff) <- input$upload_files$name    
    cat("<uploaded_files> 1\n")

    ## store files in reactive value
    files.needed = c("counts.csv","samples.csv","genes.csv","contrasts.csv")
    ff = ff[ which(names(ff) %in% files.needed) ]
    if(length(ff)>0) {
        for(i in 1:length(ff)) {
            colnames(ff[[i]]) <- gsub("[\n\t ]","_",colnames(ff[[i]]))
            rownames(ff[[i]]) <- gsub("[\n\t ]","_",rownames(ff[[i]]))
            if(names(ff)[i] %in% c("counts.csv","contrasts.csv")) {
                ff[[i]] <- as.matrix(ff[[i]])
            }
            uploaded_files[[names(ff)[i]]] <- ff[[i]]
        }
    }

    cat("<uploaded_files> 2\n")
    
    ## check dimensions
    files.uploaded <- names(uploaded_files)

    cat("<uploaded_files> 2 : files.uploaded=",files.uploaded,"\n")
    
    ##files.uploaded <- files.uploaded[match(files.needed,names(files.uploaded))]
    status = rep("please upload",4)
    names(status) = files.needed
    files.nrow = rep(NA, 4)
    files.ncol = rep(NA, 4)
    for(i in 1:4) {
        fn = files.needed[i]
        if(fn %in% files.uploaded) {
            status[i] = "OK"
            files.nrow[i] = nrow(uploaded_files[[fn]])
            files.ncol[i] = ncol(uploaded_files[[fn]])
        }
    }

    cat("<uploaded_files> 3 : status=",status,"\n")
        
    ## check files
    if(status["counts.csv"]=="OK" && status["genes.csv"]=="OK") {
        if(!all( rownames(uploaded_files[["counts.csv"]]) ==
                 rownames(uploaded_files[["genes.csv"]]) )) {
            status["counts.csv"] = "error: rownames do not match (with genes)"
            status["genes.csv"]  = "error: rownames do not match (with counts)"
        }
    }

    cat("<uploaded_files> 3b\n")
    
    if(status["counts.csv"]=="OK" && status["samples.csv"]=="OK") {
        if(!all( colnames(uploaded_files[["counts.csv"]]) ==
                 rownames(uploaded_files[["samples.csv"]]) )) {
            status["counts.csv"] = "error: colnames do not match (with samples)"
            status["samples.csv"]  = "error: rownames do not match (with counts)"
        }
    }

    cat("<uploaded_files> 3c : names(status)=",names(status),"\n")
    cat("<uploaded_files> 3c : names(uploaded_files)=",names(uploaded_files),"\n")
    
    if(status["samples.csv"]=="OK" && status["contrasts.csv"]=="OK") {
        samples1   = uploaded_files[["samples.csv"]]
        contrasts1 = uploaded_files[["contrasts.csv"]]
        has.group <- "group" %in% colnames(samples1)
        matching.group <- all(rownames(contrasts1) %in% samples1$group)
        has.group
        matching.group
        cat("<uploaded_files> 3c : has.group=",has.group,"\n")
        cat("<uploaded_files> 3c : matching.group=",matching.group,"\n")
        if(!has.group) {
            status["samples.csv"] = "error: missing 'group' column"
        }
        if(has.group && !matching.group) {
            status["contrasts.csv"] = "error: contrasts do not match groups"
        }
    }

    cat("<uploaded_files> 4\n")
        
    ## check files
    description = c(
        "Count/expression file with gene on rows, samples as columns.",
        "Samples file with samples on rows, phenotypes as columns.",
        "Gene information file with genes on rows, gene info as columns.",
        "Contrast file with conditions on rows, contrasts as columns."        
    )
    df <- data.frame( status=status, filename=files.needed,
                     description=description,
                     nrow=files.nrow, ncol=files.ncol )

    ## deselect
    ## selectRows(proxy = dataTableProxy("pgxtable"), selected=NULL)

    return(df)    
})

output$upload_status <- renderTable({
    uploadStatusTable()
})


uploadedDatasetsTable <- reactive({
    df <- PGXINFO[1:2,]
    df$description <- NULL
    df$conditions <- NULL
    df$organism <- NULL
    df$nsets <- NULL
    ##df$sharing <- "private/public"
    
    buttonInput <- function(FUN, len, id, ...) {
        inputs <- character(len)
        for (i in seq_len(len)) {
            inputs[i] <- as.character(FUN(paste0(id, i), ...))
        }
        ##inputs = paste("<div style='vertical-align: bottom;'",inputs,"</div>")
        ##inputs = paste("<div style='padding-top: 8px;'",inputs,"</div>")
        inputs
    }    
    sharing <- buttonInput(
        ##FUN = shinyWidgets::materialSwitch,
        ##FUN = shinyWidgets::switchInput,
        FUN = actionButton,
        len = nrow(df),
        id = 'uploaded_share_button_',
        ##size = "mini",
        width="80px",
        inline=TRUE,
        label = "publish",
        icon = icon("universal-access"),
        style='padding:4px; font-size:100%;'
    )
    delete <- buttonInput(
        FUN = actionButton,
        len = nrow(df),
        id = 'uploaded_delete_button_',
        label = "",
        ##size = "mini",
        width="50px",
        inline=TRUE,
        icon = icon("trash"),
        style='padding:4px; font-size:100%; background-color: #B22222;'
    )
    df <- cbind(sharing, delete, df)
    return(df)    
})

output$upload_datasetsDT <- DT::renderDataTable({
    DT::datatable(
            uploadedDatasetsTable(),
            rownames=FALSE, escape = c(-1,-2),
            options = list(
                ## dom = 'T<"clear">lfrtip',
                autoWidth = TRUE, ## scrollX=TRUE,
                selection = 'none',
                ordering = FALSE, paging=FALSE, searching=FALSE, info=FALSE,
                columnDefs = list(list(width='5%', targets = list(0,1,4,5)))                
            )
        ) ##  %>%
    ## DT::formatStyle(0, target='row', fontSize='12px', lineHeight='70%')
})

output$upload_datasets <- renderTable({
    uploadedDatasetsTable()
})


```


