Expression {data-orientation=rows}
================================================================================

Inputs {.sidebar data-width=250}
--------------------------------------------------------------------------------

<br> **Differential Expression Analysis.** Compare expression between
two conditions. Determine which genes are significantly downregulated
or overexpressed in one of the groups.

```{r}

gx_infotext =
    "The Differential Expression (DE) analysis module compares expression between two conditions (i.e. tumor versus control), which is one of the fundamental analysis in the transcriptomics data analytics workflow. For each comparison of two conditions (also called 'contrast'), the analysis identifies which genes are significantly downregulated or overexpressed in one of the groups.

<br><br>The analysis starts by selecting a comparison condition that user wants to check. There are further options to filter out some genes by functional families, logarithmic foldchange (logFC) and false discovery rate (FDR). The <strong>Plots</strong> tab shows the volcano plot and MA (an application of a Bland-Altman) plot for the chosen contrast. It also shows the so-called 'signature', i.e. the top downregulated and overexpressed genes, for that contrast. Also, the expression of the selected gene across all contrasts is summarized.

<br><br>The <strong>Top genes</strong> tab shows the average expression plots across the samples for top 10 differentially expressed genes within the selected comparison. 

<br><br>Another important feature of the platform is that it displays volcano plots for all comparison simultaneously under <strong>Volcano (all)</strong> tab. This provides users an overview of the statistics of all comparisons. By comparing the volcano plots, one can immediately see which comparison is statistically weak or strong.

<br><br>The <strong>Table</strong> tab on the bottom shows the results of the statistical tests. To increase the statistical reliability of the Omics Playground, we perform the DE analysis using four commonly accepted methods in the literature, namely, t-test (standard, Welch), limma (no trend, trend, voom), edgeR (QLF, LRT), and DESeq2 (Wald, LRT), and merge the results. For a selected comparison, the results of the selected methods are combined and reported under the table, where 'meta.q' for a gene represents the highest q value among the methods and the number of stars for a gene indicate how many methods identified significant q values (q < 0.05). The table is interactive (scrollable, clickable); users can sort genes by logFC, meta.q, or average expression in either conditions. By clicking on a gene, it is possible to see which genesets contain that gene from the geneset table located on the right , and check the differential expression status in other comparisons from the plots section.

<br><br>The <strong>Foldchange (all)</strong> tab on the bottom shows

<br><br>The <strong>FDR table</strong> tab on the bottom shows

"

```

<br><br>

```{r}
##======================================================================
## Left-side main options
##======================================================================
tipify( selectizeInput("gx_contrast", "Contrast:", choices=NULL),
        "Select a contrast of interest for the analysis.", placement="top")
tipify( selectInput("gx_features","Gene family:", choices=NULL, multiple=FALSE),
        "Choose a specific gene family for the analysis.", placement="top")

FDR.VALUES = c(1e-9,1e-6,1e-3,0.01,0.05,0.1,0.2,0.5,1)
fillRow( flex=c(1,1),
         tipify( selectInput("gx_fdr","FDR", choices=FDR.VALUES, selected=0.2),
                 "Set the false discovery rate (FDR) threshold.", placement="top"),
         tipify( selectInput("gx_lfc","logFC threshold", choices=c(0,0.2,0.5,1,2,5), selected=0.5),
                 "Set the logarithmic fold change (logFC) threshold.", placement="top")
        )


br();br();br();br();br();br();
##actionLink("gx_options", "Show options ...")
tipify( actionLink("gx_options", "Options", icon=icon("cog", lib = "glyphicon")),
       "Toggle advanced options.", placement="top")

# br();br();br();
# conditionalPanel(
#     ##"input.gx_advanced",
#     "input.gx_options % 2 == 1",
#     tagList(
#         tipify( checkboxInput('gx_top10','top 10 (table)',FALSE),
#                 "Display only top 10 differntially (positively and negatively) expressed genes under the <b>Table</b> section.", placement="bottom")
#         ##checkboxInput('gx_logscale','log scale (top genes)',TRUE),
#         ##checkboxInput('gx_ungroup','ungroup (top genes)',FALSE)
#     )
# )

gx_testmethod_text = "Select a method for the statistical test. To increase the statistical reliability of the Omics Playground, we perform the DE analysis using commonly accepted methods in the literature, including t-test (standard, Welch), limma (no trend, trend, voom), edgeR (QLF, LRT), and DESeq2 (Wald, LRT), and merge the results."

br();br();br();
GX.DEFAULTTEST="trend.limma"
GX.DEFAULTTEST=c("trend.limma","edger.qlf","deseq2.wald","edger.lrt")
if(PRO.VERSION) {
    conditionalPanel(
        "input.gx_options % 2 == 1",
        tagList( tipify( checkboxGroupInput('gx_testmethod','Statistical methods:', choices=NULL, inline=TRUE),
                         gx_testmethod_text, placement="right", options = list(container = "body"))
        )
    )
}

## update choices upon change of data set 
observe({
    ngs <- inputData()
    contr <- colnames(ngs$model.parameters$contr.matrix)
    updateSelectizeInput(session, "gx_contrast", choices=sort(contr))
    
    ##fam <- names(ngs$families)
    ##fam <- grep("^TISSUE|^COMPARTMENT|^CELLTYPE|^GOCC|^DISEASE|^CUSTOM",names(GSETS),value=TRUE)
    fam <- pgx.getFamilies(ngs,nmin=10,extended=FALSE)
    updateSelectInput(session, "gx_features",choices=fam)

    gx.methods = colnames(ngs$gx.meta$meta[[1]]$fc) ## available
    sel1 = head(intersect(GX.DEFAULTTEST,gx.methods),3) ## maximum three!!
    if(PRO.VERSION) updateCheckboxGroupInput(session, 'gx_testmethod',
                                             choices = sort(gx.methods),
                                             selected = sel1)
})

selected_gxmethods <- reactive({
    ngs <- inputData()
    gx.methods0 = colnames(ngs$gx.meta$meta[[1]]$fc)
    test = head(intersect(GX.DEFAULTTEST,gx.methods0),3) ## maximum three
    if(PRO.VERSION) test = input$gx_testmethod
    test = intersect(test,gx.methods0) ## maximum three
    test
})

```


```{r message=FALSE, warning=FALSE}
comparison=1;testmethods=c("trend.limma")

getDEGtable <- function(ngs, testmethods, comparison, add.pq) {
    ##ngs = inputData()
    if(is.null(ngs)) return(NULL)
    if(is.null(testmethods)) return(NULL)
    if(is.null(comparison)) return(NULL)
    if(length(testmethods)==0 || testmethods=="") return(NULL)
    if(length(comparison)==0  || comparison=="") return(NULL)
    
    ## build meta table
    mx = ngs$gx.meta$meta[[comparison]]
    if(is.null(mx)) return(NULL)
    mm  = colnames(unclass(mx$p))
    testmethods = intersect(mm, testmethods)
    
    mx.p  = unclass(mx$p[,testmethods,drop=FALSE]) ## get rid of AsIs
    mx.q  = unclass(mx$q[,testmethods,drop=FALSE])
    mx.fc = unclass(mx$fc[,testmethods,drop=FALSE])
    ##mx$score = mx$fc * (-log10(1e-100+mx$q) )    

    ## must recompute meta parameters (maxQ method)
    mx$meta.p = apply(mx.p,1,max,na.rm=TRUE)
    mx$meta.q = apply(mx.q,1,max,na.rm=TRUE)
    mx$meta.fx = rowMeans(mx.fc,na.rm=TRUE)  
    ##mx$meta.score = rowMeans(mx$score,na.rm=TRUE)  

    fdr = 0.05
    ##fdr = as.numeric(input$gx_fdr)
    star.symbols = sapply(1:20,function(i) paste(rep("â˜…",i),collapse=""))
    stars = c("",star.symbols)[ 1 + rowSums(mx.q <= fdr, na.rm=TRUE)]        
    
    ## recalculate group averages???
    y0 <- ngs$model.parameters$exp.matrix[,comparison]
    AveExpr1 <- rowMeans(ngs$X[rownames(mx),names(which(y0>0)),drop=FALSE])
    AveExpr0 <- rowMeans(ngs$X[rownames(mx),names(which(y0<0)),drop=FALSE])
    
    ##logFC <- unclass(ngs$gx.meta$meta[[comparison]][,"fc"])[,"trend.limma"]
    logFC <- ngs$gx.meta$meta[[comparison]][,"meta.fx"]
    ##logFC <- (AveExpr1 - AveExpr0)  ## override ??? yes: see "contrast in R" Rose Maier 2015...
    ## [hack] adjust averages to match logFC...
    mean0 <- (AveExpr0+AveExpr1)/2
    AveExpr1 <- mean0 + logFC/2
    AveExpr0 <- mean0 - logFC/2

    ##gene.annot = mx[,grep("^gene|^chr",colnames(mx)),drop=FALSE]
    aa <- intersect(c("gene_name","gene_title","chr"), colnames(ngs$genes))
    gene.annot <- ngs$genes[rownames(mx),aa]
    metaq <- mx[,c("meta.q"),drop=FALSE]    
    res = data.frame( gene.annot, logFC=logFC, stars, metaq, AveExpr0, AveExpr1, check.names=FALSE )
    
    if(add.pq) {
        ## add extra columns
        ##res <- cbind( res, q=mx$q, p=mx$p)
        colnames(mx.q) <- paste0("q.",colnames(mx.q))
        res <- cbind( res, mx.q)
    }
    rownames(res) = rownames(mx)
    return(res)
}

diffExprTable <- reactive({
    ## return the full DE table 
    ngs = inputData()
    if(is.null(ngs)) return(NULL)
    comp=1;test="trend.limma"
    comp = input$gx_contrast
    gx.methods = colnames(ngs$gx.meta$meta[[1]]$fc)
    test = head(intersect(GX.DEFAULTTEST,gx.methods),3) ## maximum three
    if(PRO.VERSION) test = input$gx_testmethod
    if(is.null(comp)) return(NULL)
    if(is.null(test)) return(NULL)
    res = getDEGtable(ngs, testmethods=test, comparison=comp, add.pq=PRO.VERSION)
    return(res)
})

sigDiffExprTable <- reactive({
    ##
    ## DE table filtered by FDR and gene family
    ##
    ##
    ngs = inputData()
    if(is.null(ngs)) return(NULL)

    gx_features=1
    gx_features = input$gx_features
    if(is.null(gx_features)) return(NULL)
    ##res = getDEGtable(ngs, testmethods="trend.limma", comparison=1,add.pq=FALSE)
    res = diffExprTable()
    if(is.null(res)) return(NULL)

    fdr=1
    fdr = as.numeric(input$gx_fdr)    
    qv.col = grep("qval|adj.p|padj|fdr|meta.q",colnames(res),ignore.case=TRUE)[1]
    fx.col = grep("mean.diff|logfc|foldchange|meta.fx",colnames(res),ignore.case=TRUE)[1]        
    pval = res[,qv.col]
    if(is.null(ngs$families)) stop("FATAL:: no families in object")
    ##psel = filterProbes(ngs$genes, ngs$families[[gx_features]] )
    psel = filterProbes(ngs$genes, GSETS[[gx_features]] )
    res = res[which(pval <= fdr & rownames(res) %in% psel),,drop=FALSE]
    dim(res)

    ## filter on fold-change    
    fx.col = grep("fx|fc|sign|NES|logFC",colnames(res))[1]
    lfc = 1
    lfc = as.numeric(input$gx_lfc)
    fx  = as.numeric(res[,fx.col])
    names(fx) = rownames(res)
    print("abs problem 1")
    res = res[which(abs(fx) >= lfc),,drop=FALSE]

    if(nrow(res)==0) {
        validate(need(nrow(res) > 0, "warning. no genes passed current filters."))
        return(NULL)
    }
    print("abs problem 2")
    res = res[order(-abs(res[,fx.col])),]

    ## just show top 10
    if(input$gx_top10) {
        fx  = as.numeric(res[,fx.col])
        names(fx) = rownames(res)
        pp <- unique(c(head(names(sort(-fx[which(fx>0)])),10),
                       head(names(sort(fx[which(fx<0)])),10)))
        res = res[pp,,drop=FALSE]
        res = res[order(-res[,fx.col]),,drop=FALSE]
    }

    ## limit number of rows
    res <- head(res, 1000)

    return(res)
})
```

Row {.tabset data-height=400}
--------------------------------------------------------------------------

### Plots 

```{r message=FALSE, warning=FALSE}

##----------------------------------------------------------------------
##                     Info messsages
##----------------------------------------------------------------------

expr_plots_volcano_text = "A volcano plot for the selected comparison under the <code>Contrast</code> section."
expr_plots_maplot_text = "An application of a Bland-Altman (MA) plot for the selected comparison under the <code>Contrast</code> section."
expr_plots_topgenesbarplot_text = "The top <em>N</em>={12} differentially (both positively and negatively) expressed gene barplot for the selected comparison under the <code>Contrast</code> section."
expr_plots_topfoldchange_text = "The fold change summary barplot across all contrasts for a gene that is selected from the differential expression table under the <code>Table</code> section."


##----------------------------------------------------------------------
##                     
##----------------------------------------------------------------------

expr_plots_volcanoFUNC <- reactive({
    
    comp1=1;fdr=0.10
    comp1 = input$gx_contrast
    if(length(comp1)==0) return(NULL)
    if(is.null(input$gx_features)) return(NULL)
    ngs <- inputData()
    if(is.null(ngs)) return(NULL)
    
    fdr = 1
    fdr = as.numeric(input$gx_fdr)
    
    res = diffExprTable()
    if(is.null(res)) return(NULL)
    lfc=1
    lfc = as.numeric(input$gx_lfc)

    fam.genes = unique(unlist(GSETS[8]))
    ##fam.genes = unique(unlist(ngs$families[input$gx_features]))
    fam.genes = unique(unlist(GSETS[input$gx_features]))
    jj <- match(toupper(fam.genes),toupper(res$gene_name))
    sel.genes <- res$gene_name[setdiff(jj,NA)]
    
    fc.genes = as.character(res[,grep("^gene$|gene_name",colnames(res))])
    qval = res[,grep("adj.P.Val|meta.q|qval|padj",colnames(res))[1]]
    qval = pmax(qval, 1e-100)
    fx = res[,grep("logFC|meta.fx|fc",colnames(res))[1]]
    sig.genes = fc.genes[which(qval <= fdr & abs(fx) > lfc )]
    sel.genes = intersect(sig.genes, sel.genes)
    xlim = c(-1,1)*max(abs(fx),na.rm=TRUE)
    ylim = c(0, max(12, 1.1*max(-log10(qval),na.rm=TRUE)))
    par(mfrow=c(1,1), mar=c(4,3,2,1.5), mgp=c(2,0.8,0), oma=c(1,0,0.5,0))
    par(mfrow=c(1,1), mar=c(6,3,2,1.5), mgp=c(2,0.8,0), oma=c(0,0,0.5,0))
    gx.volcanoPlot.XY( x=fx, pv=qval, gene=fc.genes,
                      render="canvas", n=5000, nlab=10, 
                      xlim=xlim, ylim=ylim, ## hi.col="#222222",
                      use.fdr=TRUE, p.sig=fdr, lfc=lfc,
                      cex=0.9, lab.cex=1.4, cex.main=1.0,
                      xlab="effect size (log2.FC)",
                      ylab="significance (log10.q)",
                      ## main="Volcano plot",
                      highlight=sel.genes )


})


expr_plots_volcano_module <- plotModule(
    id="expr_plots_volcano", func=expr_plots_volcanoFUNC,
    info.text = expr_plots_volcano_text, label="a",
    title = "Volcano plot",
    pdf.width=10, pdf.height=10, res=75
)

output$expr_plots_volcano     <- expr_plots_volcano_module$render
output$expr_plots_volcano_pdf <- expr_plots_volcano_module$pdf


expr_plots_maplotFUNC <- reactive({
    comp1 = input$gx_contrast
    if(length(comp1)==0) return(NULL)

    ngs <- inputData()
    if(is.null(ngs)) return(NULL)
    
    fdr=1;lfc=1
    fdr = as.numeric(input$gx_fdr)    
    lfc = as.numeric(input$gx_lfc)
    
    res = diffExprTable()
    if(is.null(res)) return(NULL)
    fc.genes = as.character(res[,grep("^gene$|gene_name",colnames(res))])
    ##pval = res$P.Value
    ##pval = res[,grep("P.Value|meta.p|pval|p.val",colnames(res))[1]]

    ## filter genes by gene family or gene set
    fam.genes = unique(unlist(ngs$families[10]))
    ##fam.genes = unique(unlist(ngs$families[input$gx_features]))
    fam.genes = unique(unlist(GSETS[input$gx_features]))
    jj <- match(toupper(fam.genes),toupper(res$gene_name))
    sel.genes <- res$gene_name[setdiff(jj,NA)]

    qval = res[,grep("adj.P.Val|meta.q|qval|padj",colnames(res))[1]]
    fx = res[,grep("logFC|meta.fx|fc",colnames(res))[1]]
    sig.genes = fc.genes[which(qval <= fdr & abs(fx) > lfc )]
    sel.genes = intersect(sig.genes, sel.genes)    
    xlim = c(-1,1)*max(abs(fx),na.rm=TRUE)
    ma = rowMeans( ngs$X[rownames(res),], na.rm=TRUE)

    par(mfrow=c(1,1), mar=c(4,3,2,1.5), mgp=c(2,0.8,0), oma=c(1,0,0.5,0))
    par(mfrow=c(1,1), mar=c(6,3,2,1.5), mgp=c(2,0.8,0), oma=c(0,0,0.5,0))
    gx.volcanoPlot.XY( x=fx, pv=qval, gene=fc.genes, lfc=lfc,
                      render="canvas", n=5000, nlab=10, 
                      xlim=xlim, ylim=c(0,15),
                      xlab="average expression (log2.CPM)",
                      ylab="effect size (log2.FC)",
                      ma_plot=TRUE, ma = ma, ## hi.col="#222222",
                      use.fdr=TRUE, p.sig=fdr, ##main=comp1,
                      ## main="MA plot",
                      cex=0.9, lab.cex=1.4, cex.main=1.0,
                      highlight=sel.genes)
})

expr_plots_maplot_module <- plotModule(
    id="expr_plots_maplot", func=expr_plots_maplotFUNC,
    info.text = expr_plots_maplot_text, label="b",
    title = "MA plot",
    pdf.width=10, pdf.height=10, res=75
)

output$expr_plots_maplot     <- expr_plots_maplot_module$render
output$expr_plots_maplot_pdf <- expr_plots_maplot_module$pdf


expr_plots_topgenesbarplotFUNC <- reactive({
    require(RColorBrewer)
    ngs = inputData()
    if(is.null(ngs)) return(NULL)
    comp1 = input$gx_contrast
    if(length(comp1)==0) return(NULL)
    
    ## get table
    ##sel.row=1;pp=rownames(ngs$X)[1]
    ##sel.row = input$expr_gx_genetable_rows_selected
    
    res = sigDiffExprTable()
    if(is.null(res)) return(NULL)

    ##fc <- res$meta.fx
    fc <- res$logFC
    names(fc) <- rownames(res)
    top.up <- head(names(sort(fc[which(fc>0)],decreasing=TRUE)),12)
    top.dn <- head(names(sort(fc[which(fc<0)],decreasing=FALSE)),12)
    fc.top <- c(fc[top.up], fc[top.dn])
    klr.pal <- brewer.pal(4,"Paired")
    klr <- c( rep(klr.pal[1],length(top.up)), rep(klr.pal[2],length(top.dn)) )
    names(fc.top) <- sub(".*:","",names(fc.top))
    
    ii <- order(fc.top)
    par(mfrow=c(1,1), mar=c(4,4,2,2)*1, mgp=c(2,0.8,0), oma=c(1,1,1,0.5)*0.2)
    par(mfrow=c(1,1), mar=c(6,3,2,1), mgp=c(2,0.8,0), oma=c(0,0,0.5,0))
    barplot(fc.top[ii], las=3, cex.names=0.8, ylab="fold change",
            col=klr[ii], ylim=c(-1.1,1.2)*max(abs(fc.top),na.rm=TRUE) )

    groups <- strsplit(comp1,split="_vs_")[[1]]
    tt <- c( paste("up in",groups[1]), paste("up in",groups[2]) )
    ##tt <- c( paste("up in",groups[1]), paste("down in",groups[1]) )
    legend("topleft", legend=tt, fill=klr.pal,cex=0.9, y.intersp=0.85, bty="n")
    ##title("top DE genes",cex.main=1)
    
})

expr_plots_topgenesbarplot_module <- plotModule(
    id="expr_plots_topgenesbarplot", func=expr_plots_topgenesbarplotFUNC,
    info.text = expr_plots_topgenesbarplot_text, label="c",
    title = "top DE genes",
    pdf.width=10, pdf.height=10, res=75
)

output$expr_plots_topgenesbarplot     <- expr_plots_topgenesbarplot_module$render
output$expr_plots_topgenesbarplot_pdf <- expr_plots_topgenesbarplot_module$pdf



expr_plots_topfoldchangeFUNC <- reactive({
    require(RColorBrewer)
    ngs = inputData()
    if(is.null(ngs)) return(NULL)
    
    ## get table
    ##sel=1;pp=rownames(ngs$X)[1]
    sel = input$expr_gx_genetable_rows_selected
    if(is.null(sel)) return(NULL)    

    res = sigDiffExprTable()
    if(is.null(res) || is.null(sel)) return(NULL)
    psel <- rownames(res)[sel]
    gene <- ngs$genes[psel,"gene_name"]
    
    ##fc <- res$meta.fx
    comp=1
    comp = input$gx_contrast
    if(is.null(comp) || length(comp)==0) return(NULL)
    fc <- sapply( ngs$gx.meta$meta, function(x) x[psel,"meta.fx"])

    top.up <- head(names(sort(fc[which(fc>0)],decreasing=TRUE)),10)
    top.dn <- head(names(sort(fc[which(fc<0)],decreasing=FALSE)),10)
    fc.top <- c(fc[top.up], fc[top.dn])
    ii <- order(fc.top)
    fc.top <- head(c(fc.top[ii], rep(NA,99)),20)
    klr.pal <- brewer.pal(4,"Paired")
    ##klr.pal <- BLUERED(16)[c(3,14)]
    klr <- klr.pal[1 + 1*(sign(fc.top)>0)]
    
    par(mfrow=c(1,1), mar=c(4,4,2,2)*1, mgp=c(2,0.8,0), oma=c(1,1,1,0.5)*0.2)
    par(mfrow=c(1,1), mar=c(8,3,3,1), mgp=c(2,0.8,0), oma=c(1,0,0,0))
    nch <- max(nchar(names(fc.top)))
    m1 <- ifelse(nch > 12, 15, 10)
    m1 <- ifelse(nch > 30, 20, m1)

    #par( mar=c(4,m1,2,0.5) )
    par( mar=c(4,m1-0.5,2,1) )
    cex1 <- 1.0
    nn <- sum(!is.na(fc.top))
    if(nn>15) cex1 <- 0.92
    barplot(fc.top, col=klr, horiz=TRUE, las=1,
            xlim=c(-1,1)*max(abs(fc.top),na.rm=TRUE),
            cex.names=cex1, xlab="fold change (log2)")
    title(gene, cex.main=1, line=-1)
    
})

expr_plots_topfoldchange_module <- plotModule(
    id="expr_plots_topfoldchange", func=expr_plots_topfoldchangeFUNC,
    info.text = expr_plots_topfoldchange_text,
    title = "Gene in contrasts", label = "d",
    pdf.width=10, pdf.height=10, res=75
)

output$expr_plots_topfoldchange     <- expr_plots_topfoldchange_module$render
output$expr_plots_topfoldchange_pdf <- expr_plots_topfoldchange_module$pdf



expr_plots_boxplotFUNC <- reactive({
    require(RColorBrewer)
    ngs = inputData()
    if(is.null(ngs)) return(NULL)
    
    ## get table
    ##sel=1;pp=rownames(ngs$X)[1]
    sel = input$expr_gx_genetable_rows_selected
    if(is.null(sel)) return(NULL)    

    res = sigDiffExprTable()
    if(is.null(res) || is.null(sel)) return(NULL)

    psel <- rownames(res)[sel]
    ##gene = sub(".*:","",rownames(res)[sel])
    gene = ngs$genes[psel,"gene_name"]
    comp = 1
    comp = input$gx_contrast
    grouped <- !input$gx_ungroup
    logscale <- input$gx_logscale
    srt <- ifelse(grouped, 0, 30)

    par(mfrow=c(1,1), mar=c(6,4,3,0)*1, mgp=c(2,0.8,0), oma=c(1,1,1,0.5)*0.2)
    par(mfrow=c(1,1), mar=c(4,3,2,1.5), mgp=c(2,0.8,0), oma=c(1,0,0.5,0))
    pgx.plotGeneExpression(ngs, gene, comp=comp, grouped=grouped,
                           max.points=2000, names=TRUE,
                           logscale=logscale, srt=srt)

    
})

expr_plots_boxplot_module <- plotModule(
    id="expr_plots_boxplot", func=expr_plots_boxplotFUNC,
    info.text = "This is an example", pdf.width=10, pdf.height=10, res=75
)

output$expr_plots_boxplot     <- expr_plots_boxplot_module$render
output$expr_plots_boxplot_pdf <- expr_plots_boxplot_module$pdf



fillRow( flex = c(1,1,1,1), #height = 370,

         fillCol(flex = c(NA,1),
                 expr_plots_volcano_module$button,
                 plotOutput("expr_plots_volcano", height = "100%", width = "100%")),
         
         fillCol(flex = c(NA,1),
                 expr_plots_maplot_module$button,
                 plotOutput("expr_plots_maplot", height = "100%", width = "100%")),
         
         fillCol(flex = c(NA,1),
                 expr_plots_topgenesbarplot_module$button,
                 plotOutput("expr_plots_topgenesbarplot", height = "100%", width = "100%")),
         
         fillCol(flex = c(NA,1),
                 expr_plots_topfoldchange_module$button,
                 plotOutput("expr_plots_topfoldchange", height = "100%", width = "100%"))
         ##plotOutput('expr_boxplot')
)

```    

> **Plots** associated with the selected comparison in the `Contrasts` section. 
> **(a)** A volcano plot.
> **(b)** An application of a Bland-Altman (MA) plot.
> **(c)** The top DEGs barplot.
> **(d)** The fold change summary barplot across all contrasts.


### Top genes

```{r message=FALSE, warning=FALSE}

expr_topgenesFUNC <- reactive({

    ngs <- inputData()
    if(is.null(ngs)) return(NULL)
    
    ##res=getDEGtable(ngs, testmethods="trend.limma", comparison=6,add.pq=FALSE)
    res <- sigDiffExprTable()
    if(is.null(res) || nrow(res)==0) return(NULL)

    fx.col = grep("fc|fx|mean.diff|logfc|foldchange",tolower(colnames(res)))[1]
    fx.col
    fx = res[,fx.col]
    names(fx) <- rownames(res)
    top.up=top.down=rownames(ngs$X)
    top.up   <- names(sort(fx[fx>0],decreasing=TRUE))
    top.down <- names(sort(fx[fx<0]))
    head(top.up)
    head(top.down)
    
    y <- ngs$samples$group
    ngrp = length(unique(y))
    las = ifelse(ngrp>3, 3, 0)
    mar1 = ifelse(las==3, 4, 3)

    comp=1;grouped=0;logscale=1
    comp = input$gx_contrast
    grouped <- !input$gx_ungroup
    logscale <- input$gx_logscale
    
    srt=30
    ylab = ifelse(logscale, "log2CPM", "CPM")
    show.names <- ifelse(!grouped & ngrp>25, FALSE, TRUE)
    ##nx = ifelse(grouped, ngrp, length(y))
    nx = ifelse(grouped, 3, length(y))
    nc = 4
    nc = 8
    if( nx <= 3) nc <- 10
    if( nx > 10) nc <- 5
    if( nx > 25) nc <- 4

    ##nc <- 10
    par(mfrow=c(2,nc), mar=c(mar1,3.5,1,1), mgp=c(2,0.8,0), oma=c(0.1,0.6,0,0.6) )
    i=1
    for(i in 1:nc) {

        if(i > length(top.up)) { frame() }
        gene = sub(".*:","",top.up[i])
        pgx.plotGeneExpression(
            ngs, gene, comp=comp, grouped=grouped, max.points=1000,
            collapse.others=1, ylab = ylab, xlab="",
            logscale=logscale, names=show.names, srt=srt, main="")
        title( gene, cex.main=1, line=-0.6)

    }

    for(i in 1:nc) {

        if(i > length(top.down)) { frame() }
        gene = sub(".*:","",top.down[i])
        pgx.plotGeneExpression(
            ngs, gene, comp=comp, grouped=grouped,  max.points=1000,
            collapse.others=TRUE, ylab = ylab, xlab="",
            logscale=logscale, names=show.names, srt=srt, main="")
        title( gene, cex.main=1, line=-0.6)
        ##qv1 = formatC(qv[gs],format="e", digits=2)
        ##legend("topright", paste("q=",qv1), bty="n",cex=1)
    }
    
})

expr_topgenes_opts = tagList(
    tipify( checkboxInput('gx_logscale','log scale',TRUE),
            "Logarithmic scale the counts (abundance levels).", placement="top", options = list(container = "body")),
    tipify( checkboxInput('gx_ungroup','ungroup',FALSE),
            "Ungroup the samples", placement="bottom")
)

expr_topgenes_text = "The <strong>top genes</strong> section shows the average expression plots across the samples for the top 10 differentially (both positively and negatively) expressed genes for the selected comparison in the <code>Contrast</code> section."

expr_topgenes_module <- plotModule(
    id="expr_topgenes", func=expr_topgenesFUNC, options=expr_topgenes_opts,
    info.text = expr_topgenes_text, pdf.width=22, pdf.height=5, res=95,
    title="Expression barplot of top 10 DE genes"
)

output$expr_topgenes     <- expr_topgenes_module$render
output$expr_topgenes_pdf <- expr_topgenes_module$pdf

fillCol(flex = c(NA,1),
                 expr_topgenes_module$button,
                 plotOutput("expr_topgenes", height = "100%", width = "100%"))

```   

> Expression barplot of the the top 10 differentially (both positively and negatively) expressed genes for the selected comparison.


### Volcano (all)
	
```{r message=FALSE, warning=FALSE, fig.width=12, fig.height=12}
graphics.off()

expr_volcano_allFUNC <- reactive({

    ngs = inputData()
    if( is.null(ngs)) return(NULL)
    
    comp = names(ngs$gx.meta$meta)
    if(length(comp)==0) return(NULL)
    if(is.null(input$gx_features)) return(NULL)
    
    fdr = 1
    fdr = as.numeric(input$gx_fdr)
    lfc = as.numeric(input$gx_lfc)

    sel.genes = GSETS[["<all>"]]
    ##sel.genes = unique(unlist(ngs$families[input$gx_features]))
    sel.genes = unique(unlist(GSETS[input$gx_features]))

    ncomp <- length(comp)
    nr <- ceiling(sqrt(ncomp/3))
    NC <- 3*nr
    if(ncomp <= (nr-1)*NC) nr <- (nr-1)
    par(mfrow=c(nr,NC), mar=c(4,4,2,2)*0)

    ng = length(comp)
    nn = c(2, max(ceiling(ng/2),5))
    ##if(ng>12) nn = c(3,8)
    par(mfrow=nn, mar=c(2,4,2.3,2)*0, mgp=c(2.6,1,0))
    n = ceiling(sqrt(ng))
    if(ng>24) {
        n = max(ceiling(ng/3),6)
        par(mfrow=c(3,n), mar=c(4,4,2,2)*0)
    } else if(FALSE && ng <= 3) {
        par(mfrow=c(1,3), mar=c(4,4,2,2)*0)
    } else {
        n = max(ceiling(ng/2),6)
        par(mfrow=c(2,n), mar=c(4,4,2,2)*0)
    }

    
    ##comp <- head(comp,75)  ## maximum 75!!!
    withProgress(message="computing volcano plots ...", value=0, {
        i=1
        for(i in 1:length(comp)) {
            
            test = GX.DEFAULTTEST
            if(PRO.VERSION) test = input$gx_testmethod
            if(is.null(test)) return(NULL)
            res = getDEGtable(ngs, testmethods=test, comparison=i, add.pq=PRO.VERSION)
            
            fc.gene = res[,grep("^gene$|^gene_name$",colnames(res))]
            ##pv.col = grep("p.val|pval|meta.p",colnames(res),ignore.case=TRUE)[1]
            qv.col = grep("qval|adj.p|padj|fdr|meta.q",colnames(res),ignore.case=TRUE)[1]
            fx.col = grep("mean.diff|logfc|foldchange|meta.fx",colnames(res),ignore.case=TRUE)[1]
            qval = res[,qv.col]
            fx = res[,fx.col]
            
            sig.genes = fc.gene[which(qval <= fdr & abs(fx) >= lfc)]
            ##genes1 = intersect(sig.genes, sel.genes)
            genes1 = sig.genes[which(toupper(sig.genes) %in% toupper(sel.genes))]
            gx.volcanoPlot.XY( x=fx, pv=qval, gene=fc.gene,
                              render="canvas", n=1000, nlab=5, 
                              xlim=NULL, ylim=c(0,15), axes=FALSE, 
                              use.fdr=TRUE, p.sig=fdr, lfc=lfc,
                              ##main=comp[i], 
                              ## ma.plot=TRUE, use.rpkm=TRUE,
                              cex=0.6, lab.cex=1.5, highlight=genes1)

            is.first = (i%%NC==1)
            last.row = ( (i-1)%/%NC == (length(comp)-1)%/%NC )
            if(is.first) axis(2, tcl=0.5, mgp=c(-2,-1.5,0))
            if(last.row) axis(1, tcl=0.5, mgp=c(-2,-1.5,0))
            box()
            legend("topright", legend=comp[i], cex=1.0,bg="white")

            incProgress( 1/length(comp) )
        }
    })  ## progress

})

expr_volcano_all_text = "Under the <strong>Volcano (all)</strong> tab, the platform displays volcano plots for all contrasts simultaneously. This provides users an overview of the statistics for all comparisons. By comparing multiple volcano plots, the user can immediately see which comparison is statistically weak or strong."

expr_volcano_all_module <- plotModule(
    id="expr_volcano_all", func=expr_volcano_allFUNC,
    info.text = expr_volcano_all_text, pdf.width=22, pdf.height=8, res=75,
    title="Volcano plots for all contrasts"
)

output$expr_volcano_all     <- expr_volcano_all_module$render
output$expr_volcano_all_pdf <- expr_volcano_all_module$pdf

fillCol(flex = c(NA,1),
                 expr_volcano_all_module$button,
                 plotOutput("expr_volcano_all", height = "100%", width = "100%"))

```

> Simultaneous visualisation of volcano plots for all contrasts.


`r if(!PRO.VERSION) {"<!---------------------------------------------"}`


### Volcano (methods)
	
```{r message=FALSE, warning=FALSE, fig.width=12, fig.height=12}

expr_volcano_methodsFUNC <- reactive({
    comp = input$gx_contrast
    if(is.null(comp)) return(NULL)
    ngs = inputData()
    if(is.null(ngs)) return(NULL)
    if(is.null(input$gx_features)) return(NULL)
    
    fdr = 1
    fdr = as.numeric(input$gx_fdr)
    lfc = as.numeric(input$gx_lfc)
    genes = NULL
    sel.genes = head(ngs$genes$gene_name,100)
    ##sel.genes = unique(unlist(ngs$families[input$gx_features]))
    sel.genes = unique(unlist(GSETS[input$gx_features]))
    
    ##methods = names(ngs$gx.meta$output)
    methods = colnames(ngs$gx.meta$meta[[1]]$fc)
    par(mfrow=c(2,6), mar=c(4,4,2,2)*0)
    if(length(methods)>12) {
        par(mfrow=c(3,8), mar=c(4,4,2,2)*0)
    }
    
    ## meta tables
    mx = ngs$gx.meta$meta[[comp]]
    fc = unclass(mx$fc)
    ##pv = unclass(mx$p)
    qv = unclass(mx$q)
    ## fc = cbind( meta=mx[,"meta.fx"], fc)
    ## qv = cbind( meta=mx[,"meta.q"], qv)

    xlim = c(-1.1,1.1)*max(abs(fc))
    xlim = 1.3*c(-1,1) * quantile(abs(fc),probs=0.999)
    fc.genes = ngs$genes[rownames(mx),"gene_name"]
    i=1
    for(i in 1:min(24,ncol(qv))) {
        fx = fc[,i]
        ## pval = pv[,i]
        qval = qv[,i]
        sig.genes = fc.genes[which(qval <= fdr & abs(fx) >= lfc)]
        ##genes1 = intersect(sig.genes, sel.genes)
        genes1 = sig.genes[which(toupper(sig.genes) %in% toupper(sel.genes))]
        gx.volcanoPlot.XY(
            x=fx, pv=qval, gene=fc.genes,
            render="canvas", n=5000, nlab=5, 
            xlim=xlim, ylim=c(0,15), axes=FALSE, 
            use.fdr=TRUE, p.sig=fdr, lfc=lfc,
            ##main=comp[i], 
            ## ma.plot=TRUE, use.rpkm=TRUE,
            cex=0.6, lab.cex=1.5, highlight=genes1)
        axis(2, tcl=0.5, mgp=c(-2,-1.5,0))
        axis(1, tcl=0.5, mgp=c(-2,-1.5,0))
        box()
        legend("topright", legend=colnames(fc)[i], cex=1.2, bg="white")
    }

})

expr_volcano_methods_text = "Under the <strong>Volcano (methods)</strong> tab, the platform displays the volcano plots of multiple methods for the selected comparison in the <code>Contrast</code> section. This provides users an overview of the statistics of all methods at the same time."

expr_volcano_methods_module <- plotModule(
    id="expr_volcano_methods", func=expr_volcano_methodsFUNC,
    info.text = expr_volcano_methods_text, pdf.width=22, pdf.height=8, res=75,
    title="Volcano plots for all methods"
)

output$expr_volcano_methods     <- expr_volcano_methods_module$render
output$expr_volcano_methods_pdf <- expr_volcano_methods_module$pdf

fillCol(flex = c(NA,1),
                 expr_volcano_methods_module$button,
                 plotOutput("expr_volcano_methods", height = "100%", width = "100%"))

```

> Simultaneous visualisation of volcano plots provided by multiple methods for the selected contrast.


`r if(!PRO.VERSION) {"------------------------------------------->"}`


Row {.tabset data-height=400}
-------------------------------------------------------------------------------------

### Table

```{r echo=FALSE, warning=FALSE, fig.height=5, fig.width=8}

expr_gx_genetableFUNC <- reactive({
    res <- sigDiffExprTable()
    if(is.null(res) || nrow(res)==0) return(NULL)
    
    fx.col = grep("fc|fx|mean.diff|logfc|foldchange",tolower(colnames(res)))[1]
    fx.col
    fx = res[,fx.col]
    ##cat("dim.rpt=",dim(rpt),"\n")
    ##cat("colnames.rpt=",colnames(rpt),"\n")
    if("gene_title" %in% colnames(res)) res$gene_title = shortstring(res$gene_title,50)
    rownames(res) = sub(".*:","",rownames(res))

    if(!PRO.VERSION) {
        kk = grep("meta.fx|meta.fc|meta.p",colnames(res),invert=TRUE)
        res <- res[,kk,drop=FALSE]
    }

    numeric.cols <- which(sapply(res, is.numeric))
    numeric.cols
   
    DT::datatable( res, rownames=FALSE,
                  class = 'compact cell-border stripe hover',                  
                  extensions = c('Buttons','Scroller'),
                  selection=list(mode='single', target='row', selected=1),
                  options=list(
                      dom = 'lfrtip',
                      pageLength = 400,
                      buttons = list('copy','csv',
                                     list(extend='pdf',
                                          pageSize="A3", orientation='landscape', 
                                          filename="diffexpr-table", title="")),
                      ##pageLength = 20,##  lengthMenu = c(20, 30, 40, 60, 100, 250),
                      scrollX = TRUE, scrollY = 300, scroller=TRUE, deferRender=TRUE
                  )  ## end of options.list 
                  ) %>%
        formatSignif(numeric.cols,4) %>%
        DT::formatStyle(0, target='row', fontSize='11px', lineHeight='70%')  %>%
            DT::formatStyle(colnames(res)[fx.col],
                            ##background = styleColorBar(c(0,3), 'lightblue'),
                            background = color_from_middle(fx, 'lightblue', '#f5aeae'),
                            backgroundSize = '98% 88%',
                            backgroundRepeat = 'no-repeat',
                            backgroundPosition = 'center')
#}, server=FALSE)
})

expr_gx_genetable_text = "The table <strong>I</strong> shows the results of the statistical tests. To increase the statistical reliability of the Omics Playground, we perform the DE analysis using four commonly accepted methods in the literature, namely, <a href='https://en.wikipedia.org/wiki/Student%27s_t-test'>t-test</a> (standard, Welch), <a href='https://www.ncbi.nlm.nih.gov/pubmed/25605792'> limma</a> (no trend, trend, voom), <a href='https://www.ncbi.nlm.nih.gov/pubmed/19910308'> edgeR</a> (QLF, LRT), and <a href='https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4302049'> DESeq2</a> (Wald, LRT), and merge the results. For a selected comparison under the <code>Contrast</code> section, the results of the selected methods are combined and reported under the table, where <code>meta.q</code> for a gene represents the highest <code>q</code> value among the methods and the number of stars for a gene indicate how many methods identified significant <code>q</code> values (<code>q < 0.05</code>). The table is interactive (scrollable, clickable); users can sort genes by <code>logFC</code>, <code>meta.q</code>, or average expression in either conditions."

expr_gx_genetable_module <- tableModule(
    id="expr_gx_genetable", func=expr_gx_genetableFUNC,
    info.text = expr_gx_genetable_text, label="I", info.text.width="450px",
    options = tipify( checkboxInput('gx_top10','top 10 genes',FALSE),
    "Display only top 10 differntially (positively and negatively) expressed genes under the <b>Table</b> section.", 
    placement="top", options = list(container = "body")),
    title="Differential expression"
)

output$expr_gx_genetable <- expr_gx_genetable_module$render

gx_related_genesets <- reactive({
    ngs <- inputData()
    res <- sigDiffExprTable()
    if(is.null(res) || nrow(res)==0) return(NULL)
    contr <- input$gx_contrast
    if(is.null(contr)) return(NULL)
        
    ## get table
    sel.row=1;
    sel.row = input$expr_gx_genetable_rows_selected
    if(is.null(sel.row)) return(NULL)
    gene0 <- rownames(res)[sel.row]
    gene1 <- sub(".*:","",gene0)

    gset <- names(which(ngs$GMT[gene1,] != 0))
    if(length(gset)==0) return(NULL)
    
    fx <- ngs$gset.meta$meta[[contr]]$meta.fx 
    names(fx) <- rownames(ngs$gset.meta$meta[[contr]])
    fx  <- round(fx[gset],digits=4)
    rho <- cor(t(ngs$gsetX[gset,]), ngs$X[gene0,])[,1]
    rho <- round(rho, digits=3)
    gset1 <- substring(gset,1,60)
    df <- data.frame(geneset=gset1, rho=rho, fx=fx, check.names=FALSE)
    df <- df[order(-abs(df$fx)),]
    return(df)
})

expr_gx_gsettableFUNC <- reactive({
    df <- gx_related_genesets()        
    if(is.null(df)) return(NULL)
    DT::datatable(df,
                  class = 'compact cell-border stripe',
                  rownames=FALSE,
                  extensions = c('Buttons','Scroller'),
                  options=list(
                      dom = 'lfrtip', 
                      buttons = list('copy','csv',
                                     list(extend='pdf', orientation='landscape', pageSize="A3",
                                          filename="gsea-genes", title="")),
                      ##pageLength = 20,##  lengthMenu = c(20, 30, 40, 60, 100, 250),
                      scrollX = TRUE, ##scrollY = TRUE,
                      scrollY = 300, scroller=TRUE, deferRender=TRUE
                  ),  ## end of options.list 
                  selection=list(mode='single', target='row', selected=1)) %>%
        ##formatSignif(1:ncol(df),4) %>%
        DT::formatStyle(0, target='row', fontSize='11px', lineHeight='70%')  %>%
            DT::formatStyle("fx", background = color_from_middle( df$fx, 'lightblue', '#f5aeae'))
#}, server=FALSE)
})

expr_gx_gsettable_text = "By clicking on a gene in the table <code>I</code>, it is possible to see which genesets contain that gene in this table, and check the differential expression status in other comparisons from the <code>Gene in contrasts</code> plot under the <code>Plots</code> section."

expr_gx_gsettable_module <- tableModule(
    id="expr_gx_gsettable", func=expr_gx_gsettableFUNC,
    info.text = expr_gx_gsettable_text, label="II",
    title="Gene sets"
    ## server=FALSE
)

output$expr_gx_gsettable <- expr_gx_gsettable_module$render


fillRow(height = 600, flex = c(2,0.1,1), 
        
        fillCol(flex = c(NA,0.01, 1),
                expr_gx_genetable_module$button,
                br(),
                dataTableOutput("expr_gx_genetable", height = "100%", width = "100%")),
        br(),
        fillCol(flex = c(NA,0.01, 1),
                expr_gx_gsettable_module$button,
                br(),
                dataTableOutput("expr_gx_gsettable", height = "100%", width = "100%"))        
        )

```

> **Tables**. **(I)** Table I reports the differential expression analysis of genes. **(II)** Table II displays the list of gene sets and further statistics regarding a gene which is selected in Table I.


### Foldchange (all)

```{r echo=FALSE, warning=FALSE, fig.height=5, fig.width=8}

expr_gx_fctableFUNC <- reactive({
    
    ngs <- inputData()
    res <- sigDiffExprTable()
    if(is.null(res) || nrow(res)==0) return(NULL)

    ##F <- sapply(ngs$gx.meta$meta, function(x) unclass(x$fc)[,"trend.limma"])
    F <- sapply(ngs$gx.meta$meta, function(x) x$meta.fx)
    rownames(F) <- rownames(ngs$gx.meta$meta[[1]])
    dim(F)
    fc.var = F[,1]**2
    if(NCOL(F)>1) {
        fc.var <- round( apply(F,1,var), digits=4)
    }

    F1 <- data.frame( gene=rownames(F), fc.var=fc.var, round(F,digits=3), check.names=FALSE)
    F1 <- F1[order(-F1$fc.var),]
    F1 <- F1[intersect(rownames(F1),rownames(res)),]  ## take intersection of current comparison
    
    DT::datatable( F1, rownames=FALSE,
                  class = 'compact cell-border stripe hover',
                  ##rownames=FALSE,
                  extensions = c('Buttons','Scroller'),
                  selection=list(mode='single', target='row', selected=c(1)),
                  options=list(
                      dom = 'lfrtip', 
                      buttons = list('copy','csv',
                                     list(extend='pdf',
                                          pageSize="A3", orientation='landscape', 
                                          filename="diffexpr-table", title="")),
                      ##pageLength = 20,##  lengthMenu = c(20, 30, 40, 60, 100, 250),
                      scrollX = TRUE, scrollY = 280, scroller=TRUE, deferRender=TRUE
                  )  ## end of options.list 
                  ) %>%
        DT::formatStyle(0, target='row', fontSize='11px', lineHeight='70%')  %>%
            DT::formatStyle( "fc.var",
                            ##background = styleColorBar(c(0,3), 'lightblue'),
                            background = color_from_middle( fc.var, 'lightblue', '#f5aeae'),
                            backgroundSize = '98% 88%', backgroundRepeat = 'no-repeat',
                            backgroundPosition = 'center')  %>%
            DT::formatStyle( colnames(F),
                            ##background = styleColorBar(c(0,3), 'lightblue'),
                            background = color_from_middle(F, 'lightblue', '#f5aeae'),
                            backgroundSize = '98% 88%', backgroundRepeat = 'no-repeat',
                            backgroundPosition = 'center')
})

expr_gx_fctable_text = "The <strong>Foldchange (all)</strong> tab reports the fold changes for all contrasts in the selected dataset."

expr_gx_fctable_module <- tableModule(
    id="expr_gx_fctable", func=expr_gx_fctableFUNC,
    info.text = expr_gx_fctable_text,
    title ="Fold changes for all contrasts"
)

output$expr_gx_fctable     <- expr_gx_fctable_module$render
output$expr_gx_fctable_csv <- expr_gx_fctable_module$csv

fillCol(flex = c(NA,0.01, 1),
        expr_gx_fctable_module$button,
        br(),
        dataTableOutput("expr_gx_fctable", height = "100%", width = "100%")) 

```

> This table reports the fold changes for all contrasts

`r if(!PRO.VERSION) {"<!---------------------------------------------"}`

### FDR table

```{r message=FALSE, warning=FALSE}
#### Number of significant genes

expr_FDR_tableFUNC <- reactive({
    
    methods = GX.DEFAULTTEST
    if(PRO.VERSION) methods = input$gx_testmethod
    ##methods = input$gx_testmethod
    if(is.null(methods)) return(NULL)

    ##comp <- input$gx_contrast
    ngs = inputData()

    kk = rownames(ngs$gx.meta$sig.counts[[1]][[1]])
    kk = intersect(methods, rownames(ngs$gx.meta$sig.counts[[1]][[1]]))
    counts.up <- ngs$gx.meta$sig.counts$up
    counts.down <- ngs$gx.meta$sig.counts$down
    counts.up <- lapply(counts.up, function(x) x[kk,,drop=FALSE])
    counts.down <- lapply(counts.down, function(x) x[kk,,drop=FALSE])
    for(i in 1:length(counts.up)) {
        rownames(counts.up[[i]]) = paste0(names(counts.up)[i],"::",rownames(counts.up[[i]]))
        rownames(counts.down[[i]]) = paste0(names(counts.down)[i],"::",rownames(counts.down[[i]]))
    }
    sig.up = do.call(rbind, counts.up)
    sig.down = do.call(rbind, counts.down)
    
    sig.up   <- sig.up[order(rownames(sig.up)),,drop=FALSE]
    sig.down <- sig.down[order(rownames(sig.down)),,drop=FALSE]    
    colnames(sig.up)[1] = paste("UP   FDR = ",colnames(sig.up)[1])
    colnames(sig.down)[1] = paste("DOWN   FDR = ",colnames(sig.down)[1])
    colnames(sig.down) = paste0("  ",colnames(sig.down))
    sigcount = cbind( sig.down, sig.up[rownames(sig.down),,drop=FALSE] )
    dim(sigcount)    
    maxsig = 0.99 * max(sigcount,na.rm=TRUE)

    contr = sub("::.*","",rownames(sigcount))
    ##contr = rownames(sigcount)
    metd  = sub(".*::","",rownames(sigcount))
    D = data.frame( method=metd, contrast=contr, sigcount, check.names=FALSE)
    
    DT::datatable( D, rownames=FALSE,
                  class = 'compact cell-border stripe hover',
                  options=list(
                      extensions = c('Buttons','Scroller'),                      
                      dom = 't',
                      pageLength = 999, ##  lengthMenu = c(20, 30, 40, 60, 100, 250),
                      scrollX = TRUE, scrollY = 280, scroller=TRUE, deferRender=TRUE
                  )  ## end of options.list 
                  ) %>%
        DT::formatStyle(0, target='row', fontSize='11px', lineHeight='70%') %>%
            DT::formatStyle(colnames(sig.up),
                            background = styleColorBar(c(0,maxsig), '#f5aeae'),
                            backgroundSize = '98% 88%',
                            backgroundRepeat = 'no-repeat',
                            backgroundPosition = 'center')  %>%
            DT::formatStyle(colnames(sig.down),
                            background = styleColorBar(c(0,maxsig), 'lightblue'),
                            backgroundSize = '98% 88%',
                            backgroundRepeat = 'no-repeat',
                            backgroundPosition = 'center')
})

expr_FDR_table_text = "This is an example"

expr_FDR_table_module <- plotModule(
    id="expr_FDR_table", func=expr_FDR_tableFUNC,
    plotlib = "datatable", info.text = expr_FDR_table_text,
    options = NULL, no.download = TRUE, server=FALSE
    ,title='Number of significant genes'
)

output$expr_FDR_table <- expr_FDR_table_module$render

fillCol(flex = c(NA,0.01, 1),
        expr_FDR_table_module$button,
        br(),
        dataTableOutput("expr_FDR_table", height = "100%", width = "100%")) 

```


`r if(!PRO.VERSION) {"------------------------------------------->"}`


